{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# CAFE\n",
    "\n",
    "\n",
    "1. Callibrate tree using the R package APE\n",
    "2. Run CAFE to identify gene family expansions and contractions. Gene families are defined as the Orthogroups foud by Orthofindr"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## APE  package in R\n",
    "\n",
    "- APE allows to callibrate a tree in multiple points.\n",
    "\n",
    "- I used it to callibrate my tree on 4 nodes:\n",
    "\n",
    "    *Blattodea + Oryhoptera. Misof node 130, 248mya\n",
    "    *Holometabola, misfo node 37, 345mya\n",
    "    *Tree root, holo+hemi,133, 390 ~mya misof node (I allowed +-5)\n",
    "    *hemiptera+thrips, misof 37, 339mya\n",
    "\n",
    "\n",
    "- I used an \"strict clock\" model  (model = \"discrete\", control = chronos.control(nb.rate.cat =1)). Which gave me the best ages on other nodes.\n",
    "\n",
    "\n",
    "- R code:\n",
    "\n",
    "-------\n",
    "\n",
    "\n",
    "```r\n",
    "\n",
    "setwd(\"Gene_Family_expansions\")\n",
    "library(\"ape\")\n",
    "library(phytools)\n",
    "InputTree<-read.tree(\"Fasttree_out_fullnames_rooted.tree\")\n",
    "\n",
    "nodestocallibrate <- c(\n",
    "  getMRCA(InputTree, tip = c(\"Gryllus_bimaculatus\",\"Blattella_germanica\") ),\n",
    "  getMRCA(InputTree, tip = c(\"Apis_mellifera\",\"Drosophila_melanogaster\") ),\n",
    "  getMRCA(InputTree, tip = c(\"Gryllus_bimaculatus\",\"Drosophila_melanogaster\") ),\n",
    "  getMRCA(InputTree, tip = c(\"Frankliniella_occidentalis\",\"Acyrthosiphon_pisum\") ) \n",
    "  \n",
    ") \n",
    "age.min <- c(\n",
    "  248,\n",
    "  345,\n",
    "  385,\n",
    "  339\n",
    ")\n",
    "age.max <- c(\n",
    "  248,\n",
    "  345,\n",
    "  395,\n",
    "  339\n",
    ")\n",
    "\n",
    "my_4points <- data.frame(nodestocallibrate, age.min, age.max) \n",
    "\n",
    "mycalibration2 <- makeChronosCalib(InputTree,node=nodestocallibrate, my_4points)\n",
    "### strict clock model:\n",
    "mytimetree_4poitnscallibrated <- chronos(InputTree,  model = \"discrete\", calibration = mycalibration2, control = chronos.control(nb.rate.cat =1) )\n",
    "mytimetree_4poitnscallibrated\n",
    "\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## APE result:\n",
    "\n",
    "**Fasttree_out_fullnames_rooted_Ape_chronos_Callirated_4points.tree**\n",
    "\n",
    "\n",
    "(((Frankliniella_occidentalis:339,((Acyrthosiphon_pisum:40.38991217,Sipha_flava:40.38991217):254.0640153,(Laodelphax_striatella:48.99958058,Nilaparvata_lugens:48.99958058):245.4543469):44.5460725):35.82192079,(Pediculus_humanus:362.4323662,(Apis_mellifera:345,(Tribolium_castaneum:314.9941411,(Drosophila_melanogaster:279.4348203,Bombyx_mori:279.4348203):35.55932079):30.00585895):17.43236624):12.38955455):13.58490598,((Blattella_germanica:112.528622,(Cryptotermes_secundus:61.00174814,Zootermopsis_nevadensis:61.00174814):51.52687381):135.471378,((Gryllus_bimaculatus:89.20882147,Laupala_kohalensis:89.20882147):126.4915381,Locusta_migratoria:215.7003595):32.29964046):140.4068268);"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Gene Families File\n",
    "\n",
    "CAFE requires a file in which for each gene family there is the number of genes in each species.\n",
    "\n",
    "Tutorial recommends: You can  remove  gene  families  with  large  variance  from  your  dataset,  but  we  found  that putting aside the gene families in which one or more species have≥ 100 gene copies doesthe tric\n",
    "\n",
    "\n",
    "**CAFE doesn't like to have long names of with _... So I I change whole names to abbreviations**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "dictionary_spp_file= {\"Api\" : \"Acyrthosiphon_pisum\",\n",
    "        \"Fca\" : \"Folsomia_candida\",\n",
    "        \"Lmi\" : \"Locusta_migratoria\",\n",
    "        \"Rma\" : \"Rhopalosiphum_maidis\",\n",
    "        \"Ago\" : \"Aphis_gossypii\",\n",
    "        \"Msa\" : \"Melanaphis_sacchari\",\n",
    "        \"Sfl\" : \"Sipha_flava\",\n",
    "        \"Bge\" : \"Blattella_germanica\",\n",
    "        \"Gbi\" : \"Gryllus_bimaculatus\",\n",
    "        \"Mpe\" : \"Myzus_persicae\",\n",
    "        \"Zne\" : \"Zootermopsis_nevadensis\",\n",
    "        \"Cle\" : \"Cimex_lectularius\",\n",
    "        \"Hha\" : \"Halyomorpha_halys\",\n",
    "        \"Nlu\" : \"Nilaparvata_lugens\",\n",
    "        \"Cse\" : \"Cryptotermes_secundus\",\n",
    "        \"Lst\" : \"Laodelphax_striatella\",\n",
    "        \"Oci\" : \"Orchesella_cincta\",\n",
    "        \"Dci\" : \"Diaphorina_citri\",\n",
    "        \"Lko\" : \"Laupala_kohalensis\",\n",
    "        \"Dme\" : \"Drosophila_melanogaster\",\n",
    "        \"Tca\" : \"Tribolium_castaneum\",\n",
    "        \"Bmo\" : \"Bombyx_mori\",             \n",
    "        \"Ame\" : \"Apis_mellifera\",\n",
    "        \"Foc\" : \"Frankliniella_occidentalis\",\n",
    "        \"Phu\" : \"Pediculus_humanus\"\n",
    "                     }\n",
    "## Short names for CAFE\n",
    "tree=\"(((Frankliniella_occidentalis:339,((Acyrthosiphon_pisum:40.38991217,Sipha_flava:40.38991217):254.0640153,(Laodelphax_striatella:48.99958058,Nilaparvata_lugens:48.99958058):245.4543469):44.5460725):35.82192079,(Pediculus_humanus:362.4323662,(Apis_mellifera:345,(Tribolium_castaneum:314.9941411,(Drosophila_melanogaster:279.4348203,Bombyx_mori:279.4348203):35.55932079):30.00585895):17.43236624):12.38955455):13.58490598,((Blattella_germanica:112.528622,(Cryptotermes_secundus:61.00174814,Zootermopsis_nevadensis:61.00174814):51.52687381):135.471378,((Gryllus_bimaculatus:89.20882147,Laupala_kohalensis:89.20882147):126.4915381,Locusta_migratoria:215.7003595):32.29964046):140.4068268);\"\n",
    "\n",
    "tree_abb=tree\n",
    "\n",
    "for abb, spp in dictionary_spp_file.items():\n",
    "    #print(abb, spp)\n",
    "    tree_abb = tree_abb.replace(spp, abb)\n",
    "print(tree_abb)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Create Lambdas tree for CAFE\n",
    "import re\n",
    "\n",
    "lambdatree=tree_abb.lower()\n",
    "\n",
    "lambdatree = re.sub(r'[0-9]', '', lambdatree)\n",
    "lambdatree = re.sub('\\.', '', lambdatree)\n",
    "lambdatree = re.sub('\\:', '', lambdatree)\n",
    "lambdatree = re.sub('\\)', ')1', lambdatree)\n",
    "lambdatree = re.sub('_', '1', lambdatree)\n",
    "lambdatree = re.sub('\\)1;', ');', lambdatree)\n",
    "print(lambdatree)\n",
    "lambdatree = re.sub(r'[a-z]+', r'1',lambdatree)\n",
    "print(\"\\n\\n\")\n",
    "print(lambdatree)\n",
    "\n"
   ]
  },
  {
   "attachments": {
    "image.png": {
     "image/png": "iVBORw0KGgoAAAANSUhEUgAAAxYAAAJSCAYAAABeAFNpAAAABHNCSVQICAgIfAhkiAAAIABJREFUeF7t3X+IXXV6P/DnaiQzE2MSZmijJSFiM/kxZko7grMsok0tY12kKOaPrlFn/bEGKuofKbit4D+yhkWoYJdqzHanJdJFohuLLBhSkxIJCYizqZFNLDapoqRmZpx1Z03Uifebe7b61TEx986cO/f8eF247M6dcz7neV7PMcl7zrlzK9XTj/AgQIAAAQIECBAgQIBAgwJ79g9HX++q6Ghvi/Ma3NfmBAgQIECAAAECBAgQ+JqAYPE1Ei8QIECAAAECBAgQINCogGDRqJjtCRAgQIAAAQIECBD4moBg8TUSLxAgQIAAAQIECBAg0KiAYNGomO0JECBAgAABAgQIEPiagGDxNRIvECBAgAABAgQIECDQqIBg0aiY7QkQIECAAAECBAgQ+JqAYPE1Ei8QIECAAAECBAgQINCogGDRqJjtCRAgQIAAAQIECBD4moBg8TUSLxAgQIDAlwXuvvvumDdvHhQCBAgQIPCNAoLFN/L4JgECBMot8Nvf/jbee++9+M1vflNuCN0TIECAwDkFBItzEtmAAAEC5RUYGxuLJUuWxHnn+euivGeBzgkQIFCfgL8p6nOyFQECBEopcMMNN8RTTz0V/f39cfz48bjqqquS26Kuv/76OHHiRHzwwQcxMDAQbW1tsXjx4lIaaZoAAQIEficgWDgTCBD4QqCrqysqlYpnCQy6u7vrOvO3b98e99xzT+zbty+efvrpuOmmm2J0dDR6enpi27ZtSei4+uqr48MPP4xjx47VtaaNCBAgQKCYAnOK2ZauCBCYjkDtH4zVanU6u9qnZAJfvjWqFkY9CBAgQICAKxbOAQIECBCoS6D226Gee+656OzsjNdffz1uvvnm5GrGrl27Yv78+W6FqkvRRgQIECiuQOX0Tyf9eLK489UZgYYEaj959kdCQ2Q2JkCAAAECpRbYs384+npXRUd7W7hiUepTQfMECBAgQIAAAQIE0hEQLNJxtAoBAgQIECBAgACBUgsIFqUev+YJECBAgAABAgQIpCMgWKTjaBUCBAgQIECAAAECpRYQLEo9fs0TIECAAAECBAgQSEdAsEjH0SoECBAgQIAAAQIESi0gWJR6/JonQIAAAQIECBAgkI6AYJGOo1UIECBAgAABAgQIlFpAsCj1+DVPgAABAgQIECBAIB0BwSIdR6sQIECAAAECBAgQKLWAYFHq8WueAAECBAgQIECAQDoCgkU6jlYhQIAAAQIECBAgUGoBwaLU49c8AQIECBAgQIAAgXQEBIt0HK1CgAABAgQIECBAoNQCgkWpx695AgQIECBAgAABAukICBbpOFqFAAECBAgQIECAQKkFBItSj1/zBAgQIECAAAECBNIRECzScbQKAQIECBAgQIAAgVILCBalHr/mCRAgQIAAAQIECKQjIFik42gVAgQIECBAgAABAqUWECxKPX7NEyBAgAABAgQIEEhHQLBIx9EqBAgQIECAAAECBEotIFiUevyaJ0CAAAECBAgQIJCOgGCRjqNVCBAgQIAAAQIECJRaQLAo9fg1T4AAAQIECBAgQCAdAcEiHUerECBAgAABAgQIECi1gGBR6vFrngABAgQIECBAgEA6AoJFOo5WIUCAAAECBAgQIFBqAcGi1OPXPAECBAgQIEDg7AK7d++OTZs2nX0D3yHwJYFK9fSDCAECBGoClUol/JHgXCBAgAABAgTqFdizfzj6eldFR3tbuGJRr5rtSiNw9OjR5B/Ytef4+Hhp+tYoAQLFFOjq6vriz7TP/2zzv7/7M75oDt3d3XWfxK+++mpyXrS1tcWNN94Yp06dip07d8Z3v/vdWLJkSfT29kbt70NXLOomteFpgTkUCBD4qsCyZcuSn9pfd911aAgQIJB7gdHRUVcicz/F9BuYP39+9PT0xKFDh2L79u3x1ltvJQf59NNPk9deeOGFeOKJJ+KGG25I/+BWLKyAKxaFHa3GCBAgQIAAAQJnFnj++efjrrvuirfffjuuvfba+OSTT5INP7+Kc+a9vErgmwVcsfhmH98lQIAAAQIECBROoHZV/jvf+U48+eSTsXz58i/6O//882PFihWxcOHC+Ld/+7f4n//5n8L1rqHmCQgWzbO1MgECBAgQIEAgkwJ//Md/HO+9995Xajt27Fj09fXFv/7rv37x+qWXXhrXXHNNJntQVPYE3AqVvZmoKAMCtUvBL730UixatCi2bt2agYqUQIAAAQIECBDItoBfN5vt+aiOwKwK1AKVXzc7q+QORqDpAv67bjqxAxAotYBfN1vq8WueAAECBAgQIECAQPoCboVK39SKBAgQIJAxAZ9Pk7GBKIcAgUIKePN2IceablO1D1eq/R50j+IL1G6Z8CBQRAGfT1PEqeqJAIGsCQgWWZtIBuvx4UoZHEqTShIsmgRrWQIECBAgUAIBt0KVYMhaJECAAAECBAgQINBsAcGi2cLWJ0CAAAECBAgQIFACAcGiBEPWIgECBAhE+HwaZwEBAgSaK+A9Fs31tToBAgQIZETAZ7RkZBDKIECgsAKuWBR2tBojQIAAAQIECBAgMHsCgsXsWTsSAQIECBAgQIAAgcIKCBaFHa3GCBAgQIAAAQIECMyegPdYzJ61IxEgQCAzAj74MjOjaHohPp+m6cQOQIDA/wkIFk4FAgQIlFDAB1+WZ+iCRXlmrVMCrRZwK1SrJ+D4BAgQIECAAAECBAogIFgUYIhaIECAAAECBAgQINBqAcGi1RNwfAIECBAgQIAAAQIFEBAsCjBELRAgQIAAAQIECBBotYBg0eoJOD4BAgQIECBAgACBAggIFgUYohYIECBAgAABAgQItFpAsGj1BByfAAECBAgQIECAQAEEBIsCDFELBAgQIECAAAECBFotIFi0egKOT4AAAQIECBAgQKAAAoJFAYaoBQIECBAgQIAAAQKtFhAsWj0BxydAgAABAgQIECBQAAHBogBD1AIBAgQIECBAgACBVgsIFq2egOMTIECAAAECBDIgcPfdd8e8efNi9+7dsWnTpgxUpIS8CQgWeZuYegkQIECAAAECKQv89re/jffeey9+85vfpLyy5cokIFiUadp6JUCAAAECBAicQWBsbCyWLFkS553nn4Zn4PFSnQLOnjqhbEaAAAECBPIo0NnZGZVKxbMEBt3d3XWfort27YpLLrkkOS82btwYN9xwQzz11FPR399f9xo2JDBVYM7UF3xNgAABAgQIFEdgZGSkOM3oJDWBI0eOxPr162NwcDBWr14dR48eTd5X8eSTTybvsfAgMB0BVyymo2YfAgQIECBAgECOBe64447YsGFDvPzyy/H9738/x50oPUsCgkWWpqEWAgQIECBAgMAsCDzwwANx2WWXxUMPPRSXX375LBzRIcogUKmefpShUT1OX6B2/6XTZPp+edrTrPM0rZnVatYz87M3AQIECPxOYM/+4ejrXRUd7W3hioWzggABAgQIECBAgACBGQsIFjMmtAABAgQIECBAgAABAoKFc4AAAQIECBAgQIAAgRkLCBYzJrQAAQJ5Faj9esXPf7//+Ph4XttQNwECBAgQyISAYJGJMSiCAIFWCCxbtiz5xQQDAwOtOLxjEiBAgACBQgkIFoUap2YIECBAgAABAgQItEZAsGiNu6MSIECAAAECBAgQKJSAYFGocWqGAAECBAgQIECAQGsEBIvWuDsqAQIZEai9efull16KRYsWxdatWzNSlTIIECBAgED+BObkr2QVEyBAID0BnyqfnqWVCBAgQKDcAq5YlHv+uidAgAABAgQIECCQioBgkQqjRQgQIECAAAECBAiUW0CwKPf8dU+AAAECBAgQIEAgFQHvsUiF0SIEiiHQ2dmZfBK1R/EFzLn4M9YhAQIEZltAsJhtcccjkGGBkZGRDFentDQFBIs0Na1FgAABAjUBt0I5DwgQIECAAAECBAgQmLGAYDFjQgsQIECAAAECBAgQICBYOAcIECBAgAABAgQIEJixgGAxY0ILECBAgAABAgQIECAgWDgHCBAgQIAAAQIECBCYsYBgMWNCCxAgQIAAAQIECBAgIFg4BwgQIECAAAECBAgQmLGAYDFjQgsQIECAAAECBAgQICBYOAcIECBAgAABAgQIEJixgGAxY0ILECBAgAABAgQIECAgWDgHCBAgQIAAAQIECBCYsYBgMWNCCxAgQIAAAQIECBAgIFg4BwgQIECAAAECJRPYvXt3bNq0qWRda7fZAoJFs4WtT4AAAQIECBAgQKAEAoJFCYasRQIECBAgQIBAPQKuZNSjZJuzCQgWZ5PxOgECBAgQKIBAV1dXVCoVzxIYdHd3N3zG7tq1Ky655JLk/Ni4cWM8+OCD8YMf/CAuvPDChteyA4E5CAgQIECAAIHiCoyOjka1Wi1ugzqbkcCRI0di/fr1MTg4GKtXr47aFYt9+/YlAcODQKMCrlg0KmZ7AgQIECBAgEBBBO64447YsGFDvPzyy/H9738/5syZE7/+9a8L0p02ZltAsJhtcccjQIAAAQIECGRE4IEHHojLLrssHnroobj88sujp6cntm3bFvPnz89IhcrIk0Dl9OVR10fzNLEW1Fq779Jp0gJ4hyTQRAH/XTcRN2NLm3XGBqIcAgUT2LN/OPp6V0VHe1u4YlGw4WqHAAECBAgQIECAQCsEBItWqDsmAQIECBAgQIAAgYIJCBYFG6h2CBAgQIAAAQIECLRCQLBohbpjEiBAgMCsChw9evSLz3EYHx+f1WM7GAECBMoiIFiUZdL6JECAQIkFli1blvwSioGBgRIraJ0AAQLNFRAsmutrdQIECBAgQIAAAQKlEBAsSjFmTRIgQIAAAQIECBBoroBgUYeve3PrQLIJAQIECBAgQIBAqQUEizrG797cOpBsQoAAgYwL1D4o7qWXXopFixbF1q1bM16t8ggQIJA/gTn5K1nFBAgQIECgcYHam7c9CBAgQKB5Aq5YNM/WygQIECBAgAABAgRKIyBYlGbUGiVAgAABAgQIECDQPAHBok5b9+bWCWUzAgQIECBAgACBUgp4j0WdY3dvbp1QNiNAIBcCnZ2dySdRexRfwJyLP2MdEsiKgGCRlUmogwABArMoMDIyMotHc6hWCggWrdR3bALlEnArVLnmrVsCBAgQIECAAAECTRGY1hWLrq6uGB0dbUpBFs2egJ92ZW8mKiJAgAABAgQIZE1gWsGiFiq85yBro2xePYJF82ytTIAAAQIECBAoioBboYoySX0QIECAAAECBAgQaKGAYNFCfIcmQIAAAQIECBAgUBQBwaIok9QHAQIECBAgQIAAgRYKCBYtxHdoAgQIECBAgAABAkURECyKMkl9ECBAgAABAgQIEGihgGDRQnyHJkCAAAECBAgQIFAUAcGiKJPUBwECBAgQIECAAIEWCggWLcR3aAIECBAgQIAAAQJFERAsijJJfRAgQIAAAQIECBBooYBg0UJ8hyZAgAABAgQIZEVg9+7dsWnTpqyUo44cCggWORyakgkQIECAAAECBAhkTUCwyNpE1EOAAAECBAgQIEAghwKCRQ6HpmQCBAgQIFCvQGdnZ1QqFc8SGHR3d9d7WsTOnTtj/fr1sXTp0ujt7Y2xsbEv9t2+fXs8+eSTyddDQ0Pxs5/9rO51bVhugTnlbl/3BAgQIECg2AIjIyPFblB30xb4+OOP44033ogf/vCHsWPHjli8ePG017IjgZqAKxbOAwIECBAgQIBACQWuvPLKmD9/fvT19cXExMQXAhdeeGEcOHAgeW3v3r0llNHydAUEi+nK2Y8AAQIECBAgUECBb3/72zE8PBy1W6tOnTpVwA611CyBSvX0o9HFa/dqTmO3Rg9j+4wImHdGBqEMAgQIECBAgEDGBPbsH46+3lXR0d7mVqiMzUY5BAgQIECAAAECBHIp4FaoXI5N0QQIECBAgAABAgSyJSBYZGseqiFAgAABAgQIECCQSwHBIpdjUzQBAgQIECBAgACBbAkIFtmah2oIECBAgAABAgQI5FJAsMjl2BRNgAABAgQIECBAIFsCgkW25qEaAgQIECBAgAABArkUECxyOTZFEyBAgAABAgQIEMiWgGCRrXmohgABAgQIECBAgEAuBQSLXI5N0QQIECBAgAABAgSyJSBYZGseqiFAgAABAgQIECCQSwHBIpdjUzQBAgQIECBAgACBbAkIFtmah2oIECBAgAABAgQI5FJAsMjl2BRNgAABAgQIECBAIFsCgkW25qEaAgQIECBAgAABArkUECxyOTZFEyBAgAABAgQIEMiWgGCRrXmohgABAgQIECBAgEAuBQSLXI5N0QQIECBAgAABAgSyJSBYZGseqiFAgAABAgQIECCQSwHBIpdjUzQBAgQIECBAgACBbAkIFtmah2oIECBAgAABAgQI5FJAsMjl2BRNgAABAgQIECBAIFsCgkW25qEaAgQIECBAgAABArkUECxyOTZFEyBAgAABAgQIEMiWgGCRrXmohgABAgQIECBAgEAuBQSLXI5N0QQIECBAgAABAgSyJSBYZGseqiFAgAABAgQIECCQSwHBIpdjUzQBAgRaJ7Bz58547LHHWleAIxMgQIBAJgUEi0yORVEECBAgQIAAAQIE8iUgWORrXqolQIAAAQIECBAgkEkBwSKTY1EUgdYIdHV1RaVS8SyBQXd3d10nWe22p+9+97uxZMmS6O3tjaNHjyb7/fKXv4ylS5cmr42NjcX27dvjySefTL43NDQUP/vZz+pa30YECBAgUByBOcVpRScECMxUYHR0NKrV6kyXsX/BBD799NM4dOhQvPDCC/HEE0/EX/zFX8THH38cb7zxRvzwhz+MHTt2RFtbW8G61g4BAgQINCrgikWjYrYnQIBAyQQ+v4r15bavvPLKmD9/fvT19cXExERceOGFceDAgeT/7927t2RC2iVAgACBmoBg4TwgQIAAgW8UOP/882PFihXx6KOPxr333nvGbb/97W/H8PBw1G6xOnXq1Bm38SIBAgQIFFugcvq2h4bve6j99GoauxVbssDdmXeBhzulNbMuz6zr7bT2Hova+yk2btxY7y62I0CAAIESCezZPxx9vauio73NFYsSzV2rBAgQIECAAAECBJom4IpF02iLs7CfYhdnlufqxKzPJeT7BAgQIECAwJcFXLFwPhAgQIAAAQIECBAgkKqAN2+nymkxAgQIECBAgAABAuUUECzKOXddEyBAgAABAgQIEEhVQLBIldNiBAgQIECAAAECBMopIFiUc+66JkCAAAECBAgQIJCqgGCRKqfFCBAgQIAAAQIECJRTQLAo59x1TYAAAQIECBAgQCBVAcEiVU6LESBAgAABAgQIECingGBRzrnrmgABAgQIECBAgECqAoJFqpwWI0CAAAECBAgQIFBOAcGinHPXNQECBAgQIECAAIFUBQSLVDktRoAAAQIECBAgQKCcAoJFOeeuawIECBAgQIAAAQKpCggWqXJajAABAgQIECBAgEA5BQSLcs5d1wQIECBAgAABAgRSFRAsUuW0GAECBAgQIECAAIFyCggW5Zy7rr9BYGhoKCqVSnR0dMQtt9wSk5OT37C1bxEgQIAAAQIECNQEBAvnAYEpAoODg1GtVuP48eMxMjISw8PDjAgQIECAAAECBM4hIFicA8i3yyswd+7cWLBgQaxevbq8CDonQIAAAQIECNQpIFjUCWWzcgmcPHky7r///nj44Ydj3rx55WpetwQIECBAgACBaQgIFtNAs0uxBcbGxuL222+P++67L3p6eordrO4IECBAgAABAikJCBYpQVqmOALPPvts1J4rV65M3sT94osvFqc5nRAgQIAAAQIEmiRQOf0m1Wqja9f+sTWN3Ro9jO0zImDeGRnELJRh1rOA7BAECBAgQKBAAnv2D0df76roaG/zW6EKNFetECBAgAABAgQIEGiZgFuhWkbvwAQIECBAgAABAgSKIyBYFGeWOiFAgAABAgQIECDQMoE5LTuyA+dGoLOzM3kTs0fxBcy5+DPWIQECBAgQaJaAYNEs2QKtW/v0aY9yCAgW5ZizLgkQIECAQDME3ArVDFVrEiBAIOMCXV1dyZVIz+IbdHd31302vvrqq8k50dbWFjfeeGOcOnUqdu7cGevWrYulS5fG2rVrY/PmzXHRRRfFrbfemqw7Pj4eAwMD0dHREWvWrInDhw/XfTwbEiBQLAHBoljz1A0BAgTqEhgdHU1+bbhn8Q3efPPNus6J2kbz589PPhh0cnIytm/fHm+99Vayb+08OXjwYFx66aXx7rvvxjvvvBNHjx5NQkUtaOzYsSNOnDiRbLNt27a6j2dDAgSKJSBYFGueuiFAgAABAtMWeP755+Ouu+6Kt99+O6699tr45JNPkrX6+/uTqxS1KxLf+ta3YsGCBbFixYqYmJhIQscjjzwSJ0+eTP7/3/3d3037+HYkQCDfAoJFvuenegIECBAgkJrAddddFz/60Y+SW55qt0PV89iwYUPs3bs3Fi5cmNxG5YpFPWq2IVBMAZ+8Xcy56orAtAR88va02HK5k1nncmyKJkCAQOYEfPJ25kaiIAIECBAgQIAAAQL5FnArVL7np3oCBAgQIECAAAECmRAQLDIxBkUQIECAAAECBAgQyLeAYJHv+ameAAECBAgQIECAQCYEBItMjEERBAgQIECAAAECBPItIFjke36qJ0CAAAECBAgQIJAJAcEiE2NQBAECBAgQIECAAIF8CwgW+Z6f6gkQIECAAAECBAhkQkCwyMQYFEGAAAECBAgQIEAg3wKCRb7np3oCBAgQIECAAAECmRAQLDIxBkUQIECAAAECBAgQyLeAYJHv+ameAAECBAgQIECAQCYEBItMjEERBAgQIECAAAECBPItIFjke36qJ0CAAAECBAgQIJAJAcEiE2NQBAECBAgQIECAAIF8CwgW+Z6f6gkQIECAAAECBAhkQkCwyMQYFEGAQCsEhoaGolKpREdHR9xyyy0xOTnZijIckwABAgQIFEJAsCjEGDVBgMB0BAYHB6Narcbx48djZGQkhoeHp7OMfQgQIECAAIHTAoKF04AAgdILzJ07NxYsWBCrV68uvQUAAgQIECAwXQHBYrpy9iNAoBACJ0+ejPvvvz8efvjhmDdvXiF60gQBAgQIEGiFgGDRCnXHJEAgEwJjY2Nx++23x3333Rc9PT2ZqEkRBAgQIEAgrwKCRV4np24CBGYs8Oyzz0btuXLlyuRN3C+++OKM17QAAQIECBAoq0Dl9BsXq402X/sLeBq7NXoY2xMgMMsC/tueZfAWHs6sW4jv0AQIECiQwJ79w9HXuyo62tu8ebtAc9UKAQIECBAgQIAAgZYJuBWqZfQOTIAAAQIECBAgQKA4AoJFcWapEwIECBAgQIAAAQItE5jTsiM7MAECmRPo7OxM3sTsUXwBcy7+jHVIgACB2RYQLGZb3PEIZFig9unTHuUQECzKMWddEiBAYDYF3Ao1m9qORYAAAQIECBAgQKCgAoJFQQerLQIECBAgQIDA2QR2794dmzZtOtu3vU5gWgKCxbTY7ESAAAECBPIh0NXVlbx3yrP4Bt3d3XWflA8++GD84Ac/iAsvvDA++OCDGBgYiLa2tli8eHG8+uqryflS+/rGG2+MU6dO1b2uDcstIFiUe/66J0CAAIGCC4yOjiYfautZfIM333yz7rO5drXi0UcfjYmJiXjqqafi6quvjg8//DCOHTsW8+fPj56enpicnIzt27fHW2+9Vfe6Niy3gGBR7vnrngABAgQIECihwJw5c+LXv/71F51/+Rc6PP/883HXXXfF22+/Hddee2188sknJRTS8nQEBIvpqNmHAAECBAgQIJBjgdoViW3btiVXJ+65557YtWtX8v9rt0Jdd9118aMf/SjWrl2b3A7lQaBegcrpS6PVejf+fLtaqp3Gbo0exvYECBAg0CQBf443CTaDy5p1BoeiJAIFEtizfzj6eldFR3tbuGJRoMFqhQABAgQIECBAgECrBASLVsk7LgECBAgQIECAAIECCQgWBRqmVggQIECAAAECBAi0SkCwaJW84xIgQIAAAQIECBAokIBgUaBhaoUAAQIECBAgQIBAqwQEi1bJOy4BAgQIECBAgACBAgkIFgUaplYIECBAgAABAgQItEpAsGiVvOMSIECAAAECBAgQKJCAYFGgYWqFAAECBAgQIECAQKsEBItWyTsuAQIECBAgQIAAgQIJCBYFGqZWCBAgQIAAAQIECLRKQLBolbzjEiBAgAABAgQIECiQgGBRoGFqhQABAgTOLDA0NBSVSiU6OjrilltuicnJyTNv6FUCBAgQmLaAYDFtOjsSIECAQF4EBgcHo1qtxvHjx2NkZCSGh4fzUro6CRAgkBsBwSI3o1IoAQIECMxUYO7cubFgwYJYvXr1TJeyPwECBAhMERAsnBIECBAgUAqBkydPxv333x8PP/xwzJs3rxQ9a5IAAQKzKSBYzKa2YxEgQIBASwTGxsbi9ttvj/vuuy96enpaUoODEiBAoOgCgkXRJ6w/AgQIEIhnn302ea5cuTJ5E/eLL75IhQABAgRSFqicfjNbtdE1a38oT2O3Rg9jewIECBBokoA/x5sEm8FlzTqDQ1ESgQIJ7Nk/HH29q6KjvS1csSjQYLVCgAABAgQIECBAoFUCgkWr5B2XAAECBAgQIECAQIEE5hSoF60QIECAQJ0CnZ2dyXsNPIovYM7Fn7EOCWRFQLDIyiTUQYAAgVkUqH1InEc5BASLcsxZlwSyIOBWqCxMQQ0ECBAgQIAAAQIEci4gWOR8gMonQIAAAQIECBAgkAUBwSILU1ADAQIECBAgQIAAgZwLCBY5H6DyCRAgQIAAAQIECNQjsHXr1qg9m/UQLJola10CBAgQIECAAAECJRIQLEo0bK0SIECAAAECBAgQaJaAYNEsWesSIECAAAECBAgQaLHA+++/H1deeWVcfPEcVNhwAAAgAElEQVTF8e///u9JNTt37ox169bF0qVLY+3atbF58+a46KKL4tZbb02+Pz4+HgMDA9HR0RFr1qyJw4cP19WFYFEXk40IECBAgAABAtkW6OrqSj740rP4Bt3d3XWfjFu2bIn169fHkSNHor29/Yv9qtVqHDx4MC699NJ4991345133omjR48moaIWNHbs2BEnTpxIttm2bVtdx/MBeXUx2YgAAQIECBAgkG2B0dHRqP1j0YNAPQL9/f3JVYraFYmVK1fGggULYsWKFTExMZGcR4888khs3Lgx5s6dW89yyTZnvWJRS0LflHjrPoINCRAgQIAAAQIECBBoicCdd94Z//Iv/xLLli1LrkDU89iwYUPs3bs3Fi5cmOSBeq9YVE4nkoajbe0A09itnj5sQ4AAAQIECKQo4O/sFDEzvpRZZ3xABS1vz/7h6OtdFR3tbWe/YlHQ3rVFgAABAgQIECBAgEATBM56K1QTjmVJAgQIECBAgAABAgQKKiBYFHSw2iJAgAABAgQIECAwmwKCxWxqOxYBAgQIECBAgACBggoIFgUdrLYIECBAgEAZBYaGhpLfYlP7YK9bbrklJicny8igZwItERAsWsLuoAQIECBAgEAzBAYHB5PfXHn8+PEYGRmJ4eHhZhzGmgQInEFAsDgDipcIECBAgACBfAvUPtSr9oFfq1evzncjqieQIwHBIkfDUioBAgQIECBwboGTJ0/G/fffHw8//HDMmzfv3DvYggCBVAQEi1QYLUKAAAECBAhkQWBsbCxuv/32uO+++6KnpycLJamBQGkEBIvSjFqjBAgQIECg+ALPPvts1J4rV65M3sT94osvFr9pHRLIiEDl9Bucqo3W4iPjGxWzPQECBAgQaI2Av7Nb496Ko5p1K9Qdc8/+4ejrXRUd7W3hioXzgQABAgQIECBAgACBGQsIFjMmtAABAgQIECBAgAABAnMQECBAgAABAsUV6OzsTN5r4FF8AXMu/oyz3qFgkfUJqY8AAQIECMxAoPYhcR7lEBAsyjHnLHfpVqgsT0dtBAgQIECAAAECBHIiIFjkZFDKJECAAAECBAgQIJBlAcEiy9NRGwECBAgQIECAAIGcCAgWORmUMgkQIECAAAECBAhkWUCwyPJ01EaAAAECBAgQIEAgJwKCRU4GpUwCBAgQIECAAAECWRYQLLI8HbURIECAAAECBAgQyImAYJGTQSmTAAECBAgQIECAQJYFBIssT0dtBAgQIECAAAECBHIiIFjkZFDKJECAAAECBAgQIJBlAcEiy9NRGwECBAgQIECAAIGcCAgWORmUMgkQIECAAAECBAhkWUCwyPJ01EaAAAECBAgQIEBgBgI7d+6Mxx57bAYr1L+rYFG/lS0JECBAgAABAgQIEDiLgGBxFhgvEyBAgAABAgQIECBQv4BgUb+VLQkQIECAAAECBAjkTuC1116LJUuWRG9vbxw9ejRqt0etW7culi5dGmvXro3NmzfHRRddFLfeemvS2/Hjx+Oqq66KefPmxfXXXx8nTpyoq2fBoi4mGxEgQIAAAQIEsi3Q2dkZlUrFswQG3d3dDZ2Mn376aRw6dCgefPDBeOKJJ5J9q9VqHDx4MC699NJ4991345133klCx/j4eDz99NNx0003xejoaPT09MS2bdvqOt6curayEQECBAgQIECAQKYFRkZGMl2f4lon8Hng/HIF/f39yVWKNWvWxMqVK2PBggWxYsWKmJiY+Eqh551X/3WI+rdsnYUjEyBAgAABAgQIECAwTYHzzz8/CQ2PPvpo3Hvvvedc5e67747nnnsualfBXn/99bj55pvPuU9tg8rpyyDVurb80ka11DON3Ro9jO0JECBAgAABAgQIEMiwwJ79w9HXuyo62tvCFYsMD0ppBAgQIECAAAECBPIiIFjkZVLqJECAAAECBAgQIJBhAcEiw8NRGgECBAgQIECAAIG8CAgWeZmUOgkQIECAAAECBAhkWECwyPBwlEaAAAECBAgQIEAgLwKCRV4mpU4CBAgQIECAAAECGRYQLDI8HKURIECAAAECBAgQyIuAYJGXSamTAAECBAgQIECAQIYFBIsMD0dpBAgQIECAAAECBPIiIFjkZVLqJECAAAECBAgQIJBhAcEiw8NRGgECBAgQIECAAIG8CAgWeZmUOgkQIECAAAECBAhkWECwyPBwlEaAAAECBAgQIEAgLwKCRV4mpU4CBAgQIECAAAECGRYQLDI8HKURIECAAAECBAgQyIuAYJGXSamTAAECBAgQIECAQIYFBIsMD0dpBAgQIECAAAECBPIiIFjkZVLqJECAAAECBAgQIJBhAcEiw8NRGgECBAgQIECAAIG8CAgWeZmUOgkQIECAAAECBAhkWECwyPBwlEaAAAECBAgQIEAgLwKCRV4mpU4CBAgQIECAAAECGRYQLDI8HKURIECAAAECBAgQyIuAYJGXSamTAAECBAgQIECAQIYFBIsMD0dpBAgQIECAAAECBPIiIFjkZVLqJECAAAECBAgQIJBhAcEiw8NRGgECBAgQIECAAIG8CAgWeZmUOgkQINACgZ07d8Zjjz3WgiM7JAECBAjkTUCwyNvE1EuAAAECBAgQIEAggwKCRQaHoiQCBAgQIECAAAECeRMQLPI2MfUSIEBglgVee+21WLJkSfT29sbRo0fj2LFjccUVV8TFF18c3/ve92Lr1q0xPj4eAwMD0dHREWvWrInDhw/PcpUOR4AAAQKtFhAsWj0BxyeQIYGurq6oVCqeJTDo7u6u+8z79NNP49ChQ/Hggw/GE088ET/5yU+SQHHkyJFob29P1tm8eXPs2LEjTpw4EQcPHoxt27bVvb4NCRAgQKAYAnOK0YYuCBBIQ2B0dDSq1WoaS1mjQAKfh81vaql23jzyyCOxcePGmDt37jdt6nsECBAgUFABVywKOlhtESBAIC2B888/P1asWBGPPvpo3HvvvXHnnXfGT3/601i2bFlyC9S8efNiw4YNsXfv3li4cGFyxcsVi7T0rUOAAIH8CFRO/5Sp4R9P1v7SmMZu+VFRKYGSCvhvu6SDn2bbH374Yaxbty4ef/zxWLVq1TRXsRsBAgQI5Flgz/7h6OtdFR3tbeGKRZ4nqXYCBAi0QODkyZPJVYlLLrkkenp6hIoWzMAhCRAgkEUB77HI4lTURIAAgQwLtLW1uWqd4fkojQABAq0ScMWiVfKOS4AAAQIECBAgQKBAAoJFgYapFQIECBAgQIAAAQKtEhAsWiXvuAQIECBAgAABAgQKJCBYFGiYWiFAgAABAgQIECDQKgHBolXyjkuAAAECBAgQIECgQAKCRYGGqRUCBAgQIECAAAECrRIQLFol77gECBAgQIAAAQIECiQgWBRomFohQIAAAQIECBAg0CoBwaJV8o5LgAABAgQIECBAoEACgkWBhqkVAgQIECBAgAABAq0SECxaJe+4BAgQIECAAAECBAokIFgUaJhaSUdg3759UalUor29PW677bb47LPP0lnYKgQIECBAgACBAgsIFgUertamJ9Df3x/VajXGxsZiYmIiXnnllektZC8CBAgQIECAQIkEBIsSDVurjQnUrlpccMEFsWjRosZ2tDUBAgQIECBAoIQCgkUJh67lcwsMDQ0lt0JNTk7G8uXLz72DLQgQIECAAAECJRcQLEp+Amj/zAKDg4Px0UcfxeLFi+OZZ54580ZeJUCAAAECBAgQ+EJAsHAyEJgisGXLluTN27VboA4cOBADAwOMCBAgQIAAAQIEziFQOf0m1eo5tvnat2v/6JrGbl9bxwsECGRLwH/b2ZqHaggQIECAQNYF9uwfjr7eVdHR3hauWGR9WuojQIAAAQIECBAgkAMBwSIHQ1IiAQIECBAgQIAAgawLzMl6geprvUBXV1eMjo62vhAVNF2gdiuUBwECBAgQIEBgOgKCxXTUSrZPLVR4T005hi5YlGPOuiRAgAABAs0QcCtUM1StSYAAAQIECBAgQKBkAoJFyQauXQIECBAgQIAAAQLNEBAsmqFqTQIECBAgQIAAAQIlExAsSjZw7RIgQIAAAQIECBBohoBg0QxVaxIgQIAAAQIECBAomYBgUbKBa5cAAQIECBAgQIBAMwQEi2aoWpMAAQIECORQ4IMPPoiBgYFoa2uLxYsXx65du+KSSy6J2q+i3rhxY4yPjyff7+joiDVr1sThw4dz2KWSCRBoloBg0SxZ6xIgQCDDArUPvqz9Y9Gz+Abd3d11n4lPPfVUXH311fHhhx/GsWPH4siRI7F+/fp444034rHHHovNmzfHjh074sSJE3Hw4MHYtm1b3WvbkACB4gv4gLziz1iHBAgQ+JqAD778GokX/k/gyx+Ueccdd8R///d/xy9+8Yt4/PHH47LLLotHHnkkuXoxd+5cZgQIEPiKgCsWTggCBAgQIEAgEbjnnnuS25/mz5+f3Ar1wAMPJGHioYceissvvzw2bNgQe/fujYULFyZXu1yxcOIQIPBlgUr19KNRktofJtPYrdHD2D4jAuadkUHMQhlmPQvIGTmEWWdkEMogQIBAzgX27B+Ovt5V0dHeFq5Y5HyYyidAgAABAgQIECCQBQHBIgtTUAMBAgQIECBAgACBnAsIFjkfoPIJECBAgAABAgQIZEFAsMjCFNRAgAABAgQIECBAIOcCgkXOB6h8AgSmL7Bv377kN9u0t7fHbbfdFp999tn0F7MnAQIECBAouYBgUfITQPsEyizQ39+f/Ia7sbGxmJiYiFdeeaXMHHonQIAAAQIzEhAsZsRnZwIEiiBQu2pxwQUXxKJFi4rQjh4IECBAgEBLBASLlrA7KAECWREYGhpKboWanJyM5cuXZ6UsdRAgQIAAgdwJCBa5G5mCCRBIU2BwcDA++uij5FOGn3nmmTSXthYBAgQIECiVgGBRqnFrlgCBLwts2bIlefN27RaoAwcOxMDAACACBAgQIEBgmgKV029crDa6b+0v4mns1uhhbJ8RAfPOyCBmoQyzngXkjBzCrDMyCGUQIEAg5wJ79g9HX++q6GhvC1cscj5M5RMgQIAAAQIECBDIgoBgkYUpqIEAAQIECBAgQIBAzgXm5Lx+5RMgkKJAZ2dn8p4Dj+ILmHPxZ6xDAgQIzLaAYDHb4o5HIMMCIyMjGa5OaWkKCBZpalqLAAECBGoCboVyHhAgQIAAAQIECBAgMGMBwWLGhBYgQIAAAQIECBAgQECwcA4QIECAAAECBAgQIDBjAcFixoQWIECAAAECBAgQIEBAsHAOECBAgAABAgQIECAwYwHBYsaEFiBAgAABAgQIECBAQLBwDhAgQIAAAQIECBAgMGMBwWLGhBYgQIAAAQIECBAgQECwcA4QIECAAAECBAgQIDBjAcFixoQWIECAAAECBAgQIEBAsHAOECBAgAABAgQIECAwYwHBYsaEFiBAgAABAgQIECBAQLBwDhAgQIAAAQIECBAgMGMBwWLGhBYgQIAAAQIECBAgQECwcA4QIECAAAECBAgQIDBjAcFixoQWIECAAAECxRB44IEHolKpJM8///M/j507d8b69etj6dKl0dvbG2NjYzE+Ph4DAwPR0dERa9asicOHDxejeV0QIDBjAcFixoQWIECAQP4EOjs7v/gH5Of/kPS/v/sHddEcuru76z5BH3/88ahWq/HKK6/Ez3/+82S/jz/+ON544434zne+Ezt27IjNmzcn/3vixIk4ePBgbNu2re71bUiAQLEF5hS7Pd0RIECAwJkERkZGzvSy10ouUAsV//Ef/xHf+ta3Yu7cuYnGlVdeGfPnz4++vr7kakVtm0ceeSQ2btz4xTYlZ9M+AQL/J+CKhVOBAAECBAgQSAT+9m//Nv70T/802tra4g//8A/PqLJhw4bYu3dvLFy4MLnq5YrFGZm8SKCUApXTP3moNtp57Q+SaezW6GFsnxEB887IIJRBgAABAgQIEMiYwJ79w9HXuyo62tvCFYuMDUc5BAgQIECAAAECBPIoIFjkcWpqJkCAAAECBAgQIJAxAcEiYwNRDgECBAgQIECAAIE8CggWeZyamgkQIECAAAECBAhkTECwqGMg+/btS37zRXt7e9x2223x2Wef1bGXTQgQIECAAAECBAiUR0CwqGPW/f39yW/Bqn3i6MTERPLBQR4ECBAgQIAAAQIECPx/AcGigbOhdtXiggsuiEWLFjWwl00JECBAgAABAgQIFF9AsKhzxkNDQ8mtUJOTk7F8+fI697IZAQIECBAgQIAAgXIICBZ1znlwcDA++uijWLx4cTzzzDN17mUzAgQIECBAgAABAuUQECzqmPOWLVuSN2/XboE6cOBADAwM1LGXTQgQIECAAAECBAiUR6By+k3J1Ubbrf0jexq7NXoY22dEwLwzMghlECBAgAABAgQyJrBn/3D09a6Kjva2cMUiY8NRDgECBAgQIECAAIE8CggWeZyamgkQIECAAAECBAhkTGDOdOrp7OxM3nPgUQ4Bsy7HnHVJgAABAgQIEJiJwLSCxcjIyEyOad+cCQgWORuYcgkQIECAAAECLRBwK1QL0B2SAAECBAgQIECAQNEEBIuiTVQ/BAgQIECAAAECBFogIFi0AN0hCRAgQIAAAQIECBRNQLAo2kT1Q4AAAQIECBAgQKAFAoJFC9AdkgABAgQIECBAgEDRBASLok1UPwQIECBAgAABAgRaICBYtADdIQkQIECAAAECBAgUTUCwKNpE9UOAAAECBAgQIECgBQKCRQvQHZIAAQIECBAgQIBA0QQEi6JNVD8ECBAgQIAAAQIEWiAgWLQA3SEJECBAgAABAgQIFE1AsCjaRPVDgAABAgQIECBAoAUCgkUL0B2SAAECBAjkRWDnzp3x2GOP5aVcdRIg0EIBwaKF+A5NgAABAgTyKLB79+7YtGlTHktXMwECTRSY08S1LU2AAAECBAgUUOCaa66J2tODAAECXxZwxcL5QIAAAQIECCQCx44diyuuuCIuvvji+N73vhdbt25NXn/ttddiyZIl0dvbG0ePHg1XLJwwBAicSUCwOJOK1wgQIFBwga6urqhUKp4lMOju7q77bP7JT36SBIojR45Ee3v7F/t9+umncejQoXjwwQfjiSeeqHs9GxIgUC4Bt0KVa966JUCAQCIwOjoa1WqVBoG6BD4PoXVtbCMCBEor4IpFaUevcQIECBAg8FWBO++8M37605/GsmXLYnx8PObNm5dscP7558eKFSvi0UcfjXvvvRcbAQIEzihQOf0TKz+yOiONFz8XqP2kymnifCBQLAH/XRdrnml38+GHH8a6devi8ccfj1WrVqW9vPUIECiQwJ79w9HXuyo62tvCFYsCDVYrBAgQIEBgJgInT55M3ndzySWXRE9Pj1AxE0z7EiihgPdYlHDoWiZAgAABAmcSaGtrc4X6TDBeI0CgLgFXLOpishEBAgQIECBAgAABAt8kIFh8k47vESBAgAABAgQIECBQl4BgUReTjQgQKKLAvn37kvvJa7+v/7bbbovPPvusiG3qiQABAgQIzIqAYDErzA5CgEAWBfr7+5P7ycfGxmJiYiJeeeWVLJapJgIECBAgkAsBwSIXY1IkAQLNFKhdtbjgggti0aJFzTyMtQkQIECAQKEFBItCj1dzBAicS2BoaCi5FWpycjKWL19+rs19nwABAgQIEDiLgGBxFhgvEyBQDoHBwcH46KOPYvHixfHMM8+Uo2ldEiBAgACBJggIFk1AtSQBAvkQ2LJlS/Lm7dotUAcOHIiBgYF8FK5KAgQIECCQQYHK6TcuVjNYl5IyJFD7h5fTJEMDUQqBFAT8d50CoiUIECBAIPbsH46+3lXR0d4Wrlg4IQgQIECAAAECBAgQmLGAYDFjQgsQIECAAAECBAgQIDAHAQECBD4X6OrqitHRUSAlEKjdCuVBgAABAgTSFBAs0tS0FoGcC9RChffT5HyIdZYvWNQJZTMCBAgQqFvArVB1U9mQAAECBAgQIECAAIGzCQgWZ5PxOgECBAgQIECAAAECdQsIFnVT2ZAAAQIECBAgQIAAgbMJCBZnk/E6AQIECBAgQIAAAQJ1CwgWdVPZkAABAgQIECBAgACBswkIFmeT8ToBAgQIECBAgAABAnULCBZ1U9mQAAECBAgQIECAAIGzCQgWZ5PxOgECBAgQIECAAAECdQsIFnVT2ZAAAQIECBAgQIAAgbMJCBZnk/E6AQIECBAgQIAAAQJ1CwgWdVPZkAABAgQIECBAgACBswkIFmeT8ToBAgQIECBAgAABAnULzKl7SxsSIECAAAEChRbYvXt37Nu3L6644or45S9/GRs3bix0v5ojQCBdAVcs0vW0GgECBAgQIECAAIFSCggWpRy7pgkQIECAAAECBAikKyBYpOtptQII1G4DqFQq0d7eHrfddlt89tlnBehKCwQIEGhc4I033ohbbrkl3n333Th+/HhcddVVMW/evLj++uvjxIkTjS9oDwIECi0gWBR6vJqbjkB/f39Uq9UYGxuLiYmJeOWVV6azjH0IZFqgs7MzCdCexTfo7u6e1rm4Y8eO+Id/+If4p3/6p/iDP/iDePrpp+Omm26K0dHR6OnpiW3btk1rXTsRIFBcAW/eLu5sdTZDgdo/uC644IJYtGjRDFeyO4HsCYyMjGSvKBVlSmDp0qXx/vvvxwcffBCLFy/+Sm3nnefnkpkalmIIZETAnwwZGYQysiUwNDSU3Ao1OTkZy5cvz1ZxqiFAgMAsCKxcuTJ+/OMfx1//9V/Hr371q7j77rvjueeei9rVrtdffz1uvvnmWajCIQgQyJNA5fQtH9U8FazW2Reo/eS+jKdJ7f7h2q9a/JM/+ZO48847Zx++BUcs66xbQO2QBAgQIECgEAJ79g9HX++q6GhvC1csCjFSTaQpsGXLluS+89otUAcOHIiBgYE0l7cWAQIECBAgQKCQAq5YFHKs6Tblp9jpemZ5NbPO8nTURoAAAQIEsifgikX2ZqIiAgQIECBAgAABArkWcCtUrseneAIECBAgQIAAAQLZEPDrZrMxh0xX8fnvu890kYpLRaB2K5QHAQIECBAgQGA6AoLFdNRKto/fd1+egQsW5Zm1TgkQIECAQNoCboVKW9R6BAgQIECAAAECBEooIFiUcOhaJkCAAAECBAgQIJC2gGCRtqj1CBAgQIAAAQIECJRQQLAo4dC1TIAAAQIECBAgQCBtAcEibVHrESBAgAABAgQIECihgGBRwqFrmQABAgQIECBAgEDaAoJF2qLWI0CAAAECBAgQIFBCAcGihEPXMgECBAgQIECAAIG0BQSLtEWtR4AAAQIECBAgQKCEAoJFCYeuZQIECBAgQIAAAQJpCwgWaYtajwABAgQIECBAgEAJBQSLEg5dywQIECBAgAABAgTSFhAs0ha1HgECBAgQIECAAIESCggWJRy6lgkQIECAAAECBAikLSBYpC1qPQIECBAgQIAAAQIlFBAsSjh0LRMgQIAAAQIECBBIW0CwSFvUegQIECBAgAABAgRKKCBYlHDoWiZAgAABAgQIECCQtoBgkbao9QgQIECAAAECBAiUUECwKOHQtUyAAAECBAgQIEAgbQHBIm1R6xEgQIAAAQIECBAooYBgUcKha5kAAQL1CuzevTs2bdpU7+a2I0CAAIESCwgWJR6+1gkQIJCWgACSlqR1CBAgkF+BOfktXeUECBAgkBWBa665JmpPDwIECBAor4ArFuWdvc4JECBQt8DOnTtj3bp1sXTp0li7dm1s3rw5Lrroorj11luTNVyxqJvShgQIECisgGBR2NFqjEDjAp2dnVGpVDxLYNDd3d3wCVKtVuPgwYNx6aWXxrvvvhvvvPNOHD16NMbHxxteyw4ECBAgUDwBt0IVb6Y6IjBtgZGRkWnva8fiC/T39ydXKdasWRMrV66MBQsWxIoVK2JiYqL4zeuQAAECBM4p4IrFOYlsQIAAAQIECBAgQIDAuQQqpy9tV8+1ke8TIECAAAECBAgQIEBgqsCe/cPR17sqOtrbwhWLqTq+JkCAAAECBAgQIECgYQHBomEyOxAgQIAAAQIECBAgMFVAsJgq4msCBAgQIECAAAECBBoWECwaJrMDAQIECBAgQIAAAQJTBQSLqSK+JkCAAAECBAgQIECgYQHBomEyOxAgQIAAAQIECBAgMFVAsJgq4msCBAgQIECAAAECBBoWECwaJrMDAQIECBAgQIAAAQJTBQSLqSK+JkCAAAECBAgQIECgYQHBomEyOxAgQIAAAQIECBAgMFVAsJgq4msCBAgQIECAAAECBBoWECwaJrMDAQIECBAgQIAAAQJTBQSLqSK+JkCAAAECBAgQIECgYQHBomEyOxAgQIAAAQIECBAgMFVAsJgq4msCBAgQIECAAAECBBoWECwaJrMDAQIECBAgQIAAAQJTBQSLqSK+JkCAAAECBAgQIECgYQHBomEyOxAgQIAAAQIECBAgMFVAsJgq4msCBAgQIECAAAECBBoWECwaJrMDAQIECBAgQIAAAQJTBQSLqSK+JkCAAAECBAgQIECgYQHBomEyOxAgQIAAAQIECBAgMFVAsJgq4msCBAgQIECAAAECBBoWECwaJrMDAQIECBAgQIAAAQJTBQSLqSK+JkCAAAECBAgQIECgYQHBomEyOxAgQIAAAQIECBAgMFVAsJgq4msCBAgQIECAAAECBBoWECwaJrMDAQIECBAgQIAAAQJTBeZMfcHXBAgQIFCfQHd3d/zXf/1XfRvbqm6B5cuXx5tvvln39jYkQIAAgWwICBbZmIMqCBDIoYB//OZwaEomQIAAgaYJuBWqabQWJkCAAAECBAgQIFAeAcGiPLPWKQECBAgQIECAAIGmCQgWTaO1MAECBAgQIECAAIHyCAgW5Zm1TgkQIECAAAECBAg0TUCwaBqthQkQIECAAAECBAiUR0CwKM+sdUqAAAECBAgQIECgadQmHYkAAAJYSURBVAKCRdNoLUyAAIHZF9ixY0fUPl+js7Mz/uZv/iaq1epXijh48GBUKpWvPP/+7/9+9gt1RAIECBAonIBgUbiRaogAgbIKnDhxIm699dZ48skn41e/+lXs2rUrXnjhha9wXH755UnYqD1PnToVtQ+j+6u/+quykumbAAECBFIUECxSxLQUAQIEWinw6quvJlcr1q5dG7/3e78X99xzT7z00ktnLekXv/hF/NEf/VEsXrz4rNv4BgECBAgQqFdAsKhXynYECBDIuMDIyEgSKD5//P7v/34cP378rFX/+Mc/TsKHBwECBAgQSENAsEhD0RoECBDIgEBXV1e8//77X1Tyv//7v1F77UyPt956K2rPP/uzPzvTt71GgAABAgQaFhAsGiazAwECBLIp0NfXF4cPH46XX345CRhPPfVUDAwMnLHYf/zHf4y77roreRO3BwECBAgQSENgThqLWIMAAQIEWi/Q0dER//zP/5zc3jQ2NhaDg4Pxl3/5l0lhCxcujEOHDiXvp6i9yfuZZ56J//zP/2x90SogQIAAgcIIVE7/ZpCv/i7CwrSmEQIECBAgQIAAAQIEmimwZ/9w9PWuio72tnArVDOlrU2AAAECBAgQIECgJAKCRUkGrU0CBAgQIECAAAECzRQQLJqpa20CBAgQIECAAAECJREQLEoyaG0SIECAAAECBAgQaKaAYNFMXWsTIECAAAECBAgQKImAYFGSQWuTAAECBAgQIECAQDMFBItm6lqbAAECBAgQIECAQEkEBIuSDFqbBAgQIECAAAECBJop8P8AGXUcNLsb7AQAAAAASUVORK5CYII="
    }
   },
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " Manually cretae lambda tree with different lambda for Holometabolans+Psocodea, Blattodea+Orthoptera and root+Hemiptea\n",
    "\n",
    "(((foc,((api,sfl)1,(lst,nlu)1)1)1,(phu,(ame,(tca,(dme,bmo)2)2)2)2)1,((bge,(cse,zne)3)3,((gbi,lko)3,lmi)3)3);\n",
    "\n",
    "![image.png](attachment:image.png)\n",
    "\n",
    "(((1,((1,1)1,(1,1)1)1)1,(2,(2,(2,(2,2)2)2)2)2)1,((3,(3,3)3)3,((3,3)3,3)3)3);\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: 'Orthofinder/Orthofinder_input_output/V2.2/OrthoFinder/Results_Jul30/Orthogroups/Orthogroups.GeneCount.tsv'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-3-4c5617d1dcb7>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     11\u001b[0m     \u001b[0mtsv_writer_largefamily\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcsv\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwriter\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlargefamiliesfile\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdelimiter\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'\\t'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     12\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 13\u001b[0;31m     \u001b[0;32mwith\u001b[0m \u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mOrthofinder_table_path\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mtsvfile\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     14\u001b[0m         \u001b[0mreader\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcsv\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreader\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtsvfile\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdelimiter\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'\\t'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     15\u001b[0m         \u001b[0;32mfor\u001b[0m \u001b[0mrow\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mreader\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: 'Orthofinder/Orthofinder_input_output/V2.2/OrthoFinder/Results_Jul30/Orthogroups/Orthogroups.GeneCount.tsv'"
     ]
    }
   ],
   "source": [
    "Orthofinder_table_path=\"Orthofinder/Orthofinder_input_output/V2.2/OrthoFinder/Results_Jul30/Orthogroups/Orthogroups.GeneCount.tsv\"\n",
    "\n",
    "import csv\n",
    "\n",
    "\n",
    "largefamiliesfile=open('OGS_Copies_V2.1_largefamilies.tsv', 'w+')\n",
    "\n",
    "numbertoolarge=0\n",
    "with open('OGS_Copies_V2.1_filtered.tsv', 'wt') as out_file:\n",
    "    tsv_writer = csv.writer(out_file, delimiter='\\t')\n",
    "    tsv_writer_largefamily = csv.writer(largefamiliesfile, delimiter='\\t')\n",
    "\n",
    "    with open(Orthofinder_table_path) as tsvfile:\n",
    "        reader = csv.reader(tsvfile, delimiter='\\t')\n",
    "        for row in reader:\n",
    "            largestinamily=0\n",
    "            Copies_family=row[:-1] ## remove last column, \"totals\"\n",
    "            if \"OG\" in Copies_family[0]: # if it is not the first line\n",
    "                for number in Copies_family[1:]:\n",
    "                    if int(number)>int(largestinamily):\n",
    "                        largestinamily=number\n",
    "                if int(largestinamily)>100:\n",
    "                    print(Copies_family, \"Largest=\", largestinamily)\n",
    "                    numbertoolarge+= 1\n",
    "                    tsv_writer_largefamily.writerow(['(null)']+row[:-1])\n",
    "                else:\n",
    "                    #tsv_writer.writerow(['(null)',\"\\t\".join(row[:-1])])\n",
    "                    tsv_writer.writerow(['(null)']+row[:-1])\n",
    "            else:\n",
    "                #print(row[:-1])#headers\n",
    "                header=\"\\t\".join(row[:-1])\n",
    "                #lowercaseheader= [item.lower() for item in row[:-1]]\n",
    "                #tsv_writer.writerow(['Desc']+row[:-1])\n",
    "                # From Whole name to abbreviation!\n",
    "                for abb, spp in dictionary_spp_file.items():\n",
    "                    header = header.replace(spp, abb)\n",
    "                header2=\"\\t\".join(['Desc',header]).split(\"\\t\")\n",
    "                print(header2)   \n",
    "                tsv_writer.writerow(header2)\n",
    "                tsv_writer_largefamily.writerow(header2)\n",
    "\n",
    "\n",
    "print(\"Number of families  values>100 =\", numbertoolarge, \" Families filtered\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\\t\".join(['Desc',header])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Prepare CAFE\n",
    "\n",
    "\n",
    "I need to prepare the CAFE file with all the scripts\n",
    "\n",
    "CAFE script:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## CAFE with same lambda\n",
    "\n",
    "### *CAFE_v2.2.4callibrations_run0*\n",
    "load -t == Threads\n",
    "lambda  -t  == difefernt rates birth and death\n",
    "\n",
    "```\n",
    "#!cafe\n",
    "load -i OGS_Copies_V2.1_filtered.tsv -t 8 -l reports/log_V2.2.4callibrations.run0.txt\n",
    "tree (((Foc:339,((Api:40.38991217,Sfl:40.38991217):254.0640153,(Lst:48.99958058,Nlu:48.99958058):245.4543469):44.5460725):35.82192079,(Phu:362.4323662,(Ame:345,(Tca:314.9941411,(Dme:279.4348203,Bmo:279.4348203):35.55932079):30.00585895):17.43236624):12.38955455):13.58490598,((Bge:112.528622,(Cse:61.00174814,Zne:61.00174814):51.52687381):135.471378,((Gbi:89.20882147,Lko:89.20882147):126.4915381,Lmi:215.7003595):32.29964046):140.4068268);\n",
    "\n",
    "lambda -s -t  (((1,((1,1)1,(1,1)1)1)1,(1,(1,(1,(1,1)1)1)1)1)1,((1,(1,1)1)1,((1,1)1,1)1)1);\n",
    "\n",
    "report reports/report_V2.2.4callibrations.run0\n",
    "```\n",
    "\n",
    "### Run CAFE\n",
    "\n",
    "```\n",
    "~/data_disk/Software/CAFE/release/cafe CAFE_v2.2.4callibrations_run0\n",
    "\n",
    "```\n",
    "\n",
    "### Lambda value\n",
    "\n",
    "From: reports/report_V2.2.4callibrations.run0.cafe\n",
    "\n",
    "Lambda:\t0.00132125\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## CAFE with different lambdas\n",
    "\n",
    "### *CAFE_v2.2.4callibrations_run1*\n",
    "\n",
    "load -t == Threads\n",
    "lambda  -t  == difefernt rates birth and death\n",
    "\n",
    "3 different lambdas:  This tells us that the lineage leading to (and including) Orthopterans+Blattodea  MIGHT have  different  gene  family  evolution  rates than Hemipters, and that Holometabolans+Psocodea might still have different  rates.\n",
    "\n",
    "```\n",
    "#!cafe\n",
    "load -i OGS_Copies_V2.1_filtered.tsv -t 8 -l reports/log_V2.2.4callibrations.run1.txt\n",
    "tree (((Foc:339,((Api:40.38991217,Sfl:40.38991217):254.0640153,(Lst:48.99958058,Nlu:48.99958058):245.4543469):44.5460725):35.82192079,(Phu:362.4323662,(Ame:345,(Tca:314.9941411,(Dme:279.4348203,Bmo:279.4348203):35.55932079):30.00585895):17.43236624):12.38955455):13.58490598,((Bge:112.528622,(Cse:61.00174814,Zne:61.00174814):51.52687381):135.471378,((Gbi:89.20882147,Lko:89.20882147):126.4915381,Lmi:215.7003595):32.29964046):140.4068268);\n",
    "\n",
    "lambda -s -t  (((1,((1,1)1,(1,1)1)1)1,(2,(2,(2,(2,2)2)2)2)2)1,((3,(3,3)3)3,((3,3)3,3)3)3);\n",
    "\n",
    "report reports/report_V2.2.4callibrations.run1\n",
    "```\n",
    "\n",
    "\n",
    "\n",
    "### Run CAFE\n",
    "\n",
    "```\n",
    "~/data_disk/Software/CAFE/release/cafe CAFE_v2.2.4callibrations_run1\n",
    "\n",
    "```\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## RESULTS run1\n",
    "\n",
    "Process results \n",
    "\n",
    "```\n",
    "# We provide a convenient script that summarizes the output described above into tables.\n",
    "# In order to run it, just enter the following command on your shell:\n",
    "```\n",
    "\n",
    "```\n",
    "python cafetutorial_report_analysis.py -i reports/report_V2.2.4callibrations.run1.cafe -o reports/summary_V2.2.4callibrations.run1\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# Draw tree\n",
    "\n",
    "usedtree=\"(((Foc:339,((Api:40.38991217,Sfl:40.38991217):254.0640153,(Lst:48.99958058,Nlu:48.99958058):245.4543469):44.5460725):35.82192079,(Phu:362.4323662,(Ame:345,(Tca:314.9941411,(Dme:279.4348203,Bmo:279.4348203):35.55932079):30.00585895):17.43236624):12.38955455):13.58490598,((Bge:112.528622,(Cse:61.00174814,Zne:61.00174814):51.52687381):135.471378,((Gbi:89.20882147,Lko:89.20882147):126.4915381,Lmi:215.7003595):32.29964046):140.4068268)\"\n",
    "cafenodetree='(((Foc<0>,((Api<2>,Sfl<4>)<3>,(Lst<6>,Nlu<8>)<7>)<5>)<1>,(Phu<10>,(Ame<12>,(Tca<14>,(Dme<16>,Bmo<18>)<17>)<15>)<13>)<11>)<9>,((Bge<20>,(Cse<22>,Zne<24>)<23>)<21>,((Gbi<26>,Lko<28>)<27>,Lmi<30>)<29>)<25>)<19>'\n",
    "\n",
    "# rapid evolving\n",
    "%run cafetutorial_draw_tree.py -i reports/summary_V2.2.4callibrations.run1_node.txt  -t $usedtree -d $cafenodetree -o reports/summary_v2.2.4callibrations_run1_tree_rapid.png -y Rapid\n",
    "# expansions\n",
    "%run cafetutorial_draw_tree.py -i reports/summary_V2.2.4callibrations.run1_node.txt  -t $usedtree -d $cafenodetree -o reports/summary_v2.2.4callibrations_run1_tree_Expansions.png -y Expansions\n",
    "#Contractions\n",
    "%run cafetutorial_draw_tree.py -i reports/summary_V2.2.4callibrations.run1_node.txt  -t $usedtree -d $cafenodetree -o reports/summary_v2.2.4callibrations_run1_tree_Contractions.png -y Contractions\n",
    "\n",
    "## Ignore warnings"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### CAFE Comparing models with onevsmultipleλ\n",
    "\n",
    "*CAFE_v2.2.4callibrations_run_lh*\n",
    "\n",
    "Tha Lambda value calculated from tree with homogeneous lambda=  0.00132125\n",
    "\n",
    "\n",
    "```\n",
    "#!cafe\n",
    "load -i OGS_Copies_V2.1_filtered.tsv -t 36 -l reports/log_V2.2.4callibrations.run_lh.txt\n",
    "\n",
    "tree (((Foc:339,((Api:40.38991217,Sfl:40.38991217):254.0640153,(Lst:48.99958058,Nlu:48.99958058):245.4543469):44.5460725):35.82192079,(Phu:362.4323662,(Ame:345,(Tca:314.9941411,(Dme:279.4348203,Bmo:279.4348203):35.55932079):30.00585895):17.43236624):12.38955455):13.58490598,((Bge:112.528622,(Cse:61.00174814,Zne:61.00174814):51.52687381):135.471378,((Gbi:89.20882147,Lko:89.20882147):126.4915381,Lmi:215.7003595):32.29964046):140.4068268);\n",
    "\n",
    "\n",
    "lambda  -l 0.00132125\n",
    "\n",
    "genfamily  lh_dir/rnd -t 100\n",
    "\n",
    "lhtest  -d lh_dir  -t (((1,((1,1)1,(1,1)1)1)1,(2,(2,(2,(2,2)2)2)2)2)1,((3,(3,3)3)3,((3,3)3,3)3)3) -l 0.00132125  -o reports/lhtest_result.txt\n",
    "\n",
    "\n",
    "```\n",
    "\n",
    "\n",
    "\n",
    "## Run CAFE\n",
    "\n",
    "```\n",
    "~/data_disk/Software/CAFE/release/cafe CAFE_v2.2.4callibrations_run_lh\n",
    "\n",
    "```\n",
    "\n",
    "From tutorial:\n",
    "\n",
    "*Here, the genfamily command simulates the datasets (in the example above, we areasking  for  100  simulations  with-t 100).   It  estimates λ from  the  observed  data  to simulate gene families.  Then the likelihoods of the two competing models are calculatedwith the lhtest function, which takes the multi-λ tree structure, and the estimated λ value using the global-λ model*\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## RESULTS from Likelihood test\n",
    "\n",
    "From file reports/log_V2.2.4callibrations.run0.txt (log-likelihoods ar ethe engatove of  the score): -232998.457504\n",
    "From file reports/log_V2.2.4callibrations.run1.txt (log-likelihoods ar ethe engatove of  the score) -229913.526930\n",
    "\n",
    "```\n",
    "cut -f 2,4 reports/lhtest_result.txt > reports/run4_lk_diffs.txt \n",
    "\n",
    "Rscript  lhtest.R reports/run4_lk_diffs.txt \n",
    "```\n",
    "\n",
    "From tutorial\n",
    "\n",
    "*The  numbers −162576.606204  and −149055.330013  are  the  log-likelihoods  of  theg global-λ and multi-λ models from reports /log_run1.txt and reports/log_run3.txt, respectively  (the  negative  log-likelihoods  are  given  in  these  two  files).   Running  the command  above  creates  a  histogram  with  the  null  distribution  from  the  simulations(reports/lk_null.pdf). Note that the observed likelihood ratio (2×(lnLglobal−lnLmulti)) would fall on the far left tail of the null distribution, yielding a very small p-value, and meaning that thethe probability of a multi-λmodel fitting better than a global-λmodel by chance is verysmall*\n",
    "\n",
    "\n",
    "\n",
    "The likelihood ratio test (LRT) gives a value of \"-6169.861\"\n",
    "\n",
    "2*(-232998.457504-(-229913.526930)) = -6169.861\n",
    "\n",
    "\n",
    "The p value would be the number of times that the random distributions got a LRT lower than -6169.861 divided by 100 tests.\n",
    "\n",
    "If the 3 of the 100 tests have a LTR < -6169.861 , pvalue=3/100=0.03\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## CAFE check Errors\n",
    "\n",
    "\n",
    "Since the multi lamda assumptoion was better than the global lambda, I proceed with "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Calculate error\n",
    "\n",
    "***The Expansion/Contraction trees*** show ALL teh expanded/contracted famileis, NOT only the significant. In the file *reports/FILE_pub.txt* there is the number of significant in prentesis.\n",
    "\n",
    "\n",
    "The P-value sed is  \"P-value: 0.01\" as pointed out at *reports/log_FILE.txt* .\n",
    "\n",
    "\n",
    "CAFE can correct by errors on genome assembly/annotation\n",
    "\n",
    "Re-estimate lambdas using this errorr model (MUST be ran with same lambda for whole tree, after getting the error, will run CAFE again with different lambdas but considering the error):\n",
    "\n",
    "File: CAFE_v2.2.4_calculaterror\n",
    "\n",
    "```\n",
    "#!cafe\n",
    "\n",
    "load -i OGS_Copies_V2.1_filtered.tsv -t 8 -l reports/log_V2.2.4calculaterror.run1.txt\n",
    "tree (((Foc:339,((Api:40.38991217,Sfl:40.38991217):254.0640153,(Lst:48.99958058,Nlu:48.99958058):245.4543469):44.5460725):35.82192079,(Phu:362.4323662,(Ame:345,(Tca:314.9941411,(Dme:279.4348203,Bmo:279.4348203):35.55932079):30.00585895):17.43236624):12.38955455):13.58490598,((Bge:112.528622,(Cse:61.00174814,Zne:61.00174814):51.52687381):135.471378,((Gbi:89.20882147,Lko:89.20882147):126.4915381,Lmi:215.7003595):32.29964046):140.4068268);\n",
    "\n",
    "lambda -s -t  (((1,((1,1)1,(1,1)1)1)1,(1,(1,(1,(1,1)1)1)1)1)1,((1,(1,1)1)1,((1,1)1,1)1)1);\n",
    "\n",
    "report reports/report_V2.2.4calculaterror.run1\n",
    "\n",
    "```\n",
    "\n",
    "Calculate Error model:\n",
    "\n",
    "```\n",
    "python  caferror.py -i CAFE_v2.2.4_calculaterror -v 0 -f 1\n",
    "```\n",
    "\n",
    "End of log file: caferror_1/caferrorLog.txt   (didnt use the reports/log_V2.2.4calculaterror* files, because obveis tha -d parameter from caferror which i left empty and by default created caferror_1)\n",
    "\n",
    "```\n",
    "# =======================================================================\n",
    "# ************************************\n",
    "# Score with no errormodel:        232998.456131\n",
    "# Lambda with no errormodel:       0.00132125965133\n",
    "# ************************************\n",
    "# Global Error Estimation:         0.021337890625\n",
    "# Score with global errormodel:    232459.837134\n",
    "# Lambda with global errormodel:   0.00126686991522\n",
    "# ************************************\n",
    "# =======================================================================\n",
    "\n",
    "```\n",
    "\n",
    "File: caferror_1/caferror_default_output.txt\n",
    "\n",
    "```\n",
    "ErrorModel\tScore\n",
    "0.4\t232998.456131\n",
    "0.4\t258282.847854\n",
    "0.2\t240596.103286\n",
    "0.1\t234642.056371\n",
    "0.05\t232846.652157\n",
    "0.025\t232467.682988\n",
    "0.0125\t232519.290722\n",
    "0.03125\t232514.448907\n",
    "0.021875\t232459.973763\n",
    "0.0171875\t232472.003520\n",
    "0.04375\t232708.562177\n",
    "0.03046875\t232506.518755\n",
    "0.023828125\t232463.438689\n",
    "0.0205078125\t232460.386618\n",
    "0.02216796875\t232460.223555\n",
    "0.021337890625\t232459.837134\n",
    "0.0209228515625\t232459.978250\n",
    "```\n",
    "\n",
    "\n",
    "Best error model:  0.021337890625\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "### RUN CAFE correcting for genome qulaity\n",
    "\n",
    "File: CAFE_v2.2.4_corrected\n",
    "\n",
    "\n",
    "```\n",
    "#!cafe\n",
    "load -i OGS_Copies_V2.1_filtered.tsv -t 8 -l reports/log_V2.2.4_corrected.txt\n",
    "tree (((Foc:339,((Api:40.38991217,Sfl:40.38991217):254.0640153,(Lst:48.99958058,Nlu:48.99958058):245.4543469):44.5460725):35.82192079,(Phu:362.4323662,(Ame:345,(Tca:314.9941411,(Dme:279.4348203,Bmo:279.4348203):35.55932079):30.00585895):17.43236624):12.38955455):13.58490598,((Bge:112.528622,(Cse:61.00174814,Zne:61.00174814):51.52687381):135.471378,((Gbi:89.20882147,Lko:89.20882147):126.4915381,Lmi:215.7003595):32.29964046):140.4068268);\n",
    "\n",
    "lambda -s -t  (((1,((1,1)1,(1,1)1)1)1,(2,(2,(2,(2,2)2)2)2)2)1,((3,(3,3)3)3,((3,3)3,3)3)3);\n",
    "\n",
    "errormodel  -model caferror_1/cafe_errormodel_0.021337890625.txt  -all\n",
    "\n",
    "\n",
    "report reports/log_V2.2.4_corrected\n",
    "\n",
    "\n",
    "```\n",
    "\n",
    "Run:\n",
    "\n",
    "```\n",
    "~/data_disk/Software/CAFE/release/cafe CAFE_v2.2.4_corrected\n",
    "\n",
    "```\n",
    "\n",
    "### RESULTS after Correcting by genome quality\n",
    "\n",
    "Process results \n",
    "\n",
    "```\n",
    "python  cafetutorial_report_analysis.py -i reports/log_V2.2.4_corrected.cafe -o reports/summary_V2.2.4_corrected\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Draw tree\n",
    "\n",
    "usedtree=\"(((Foc:339,((Api:40.38991217,Sfl:40.38991217):254.0640153,(Lst:48.99958058,Nlu:48.99958058):245.4543469):44.5460725):35.82192079,(Phu:362.4323662,(Ame:345,(Tca:314.9941411,(Dme:279.4348203,Bmo:279.4348203):35.55932079):30.00585895):17.43236624):12.38955455):13.58490598,((Bge:112.528622,(Cse:61.00174814,Zne:61.00174814):51.52687381):135.471378,((Gbi:89.20882147,Lko:89.20882147):126.4915381,Lmi:215.7003595):32.29964046):140.4068268)\"\n",
    "cafenodetree='(((Foc<0>,((Api<2>,Sfl<4>)<3>,(Lst<6>,Nlu<8>)<7>)<5>)<1>,(Phu<10>,(Ame<12>,(Tca<14>,(Dme<16>,Bmo<18>)<17>)<15>)<13>)<11>)<9>,((Bge<20>,(Cse<22>,Zne<24>)<23>)<21>,((Gbi<26>,Lko<28>)<27>,Lmi<30>)<29>)<25>)<19>'\n",
    "\n",
    "# rapid evolving\n",
    "%run cafetutorial_draw_tree.py -i reports/summary_V2.2.4_corrected_node.txt -t $usedtree -d $cafenodetree -o reports/summary_V2.2.4_corrected_Rapid.png -y Rapid\n",
    "# expansions\n",
    "%run cafetutorial_draw_tree.py -i reports/summary_V2.2.4_corrected_node.txt -t $usedtree -d $cafenodetree -o reports/summary_V2.2.4_corrected_Expanisons.png -y Expansions\n",
    "# #Contractions\n",
    "%run cafetutorial_draw_tree.py -i reports/summary_V2.2.4_corrected_node.txt -t $usedtree -d $cafenodetree -o reports/summary_V2.2.4_corrected_Contractions.png -y Contractions\n",
    "\n",
    "## Ignore warnings"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## CAFE Large families:\n",
    "\n",
    "CAFE has issues compiuting lambdas when there are large families (>100 genes). They recomment tho compute those at the end using the previously computed lambdas:\n",
    "\n",
    "File with gene families with > 100 genes: OGS_Copies_V2.1_largefamilies.tsv\n",
    "\n",
    "Lambdas from: reports/log_V2.2.4_corrected.cafe\n",
    "\n",
    "\n",
    "File: CAFE_v2.2.4_largefamilies\n",
    "\n",
    "```\n",
    "#!cafe\n",
    "load -i OGS_Copies_V2.1_largefamilies.tsv -t 8 -l reports/log_V2.2.4_largefamilies.txt\n",
    "tree (((Foc:339,((Api:40.38991217,Sfl:40.38991217):254.0640153,(Lst:48.99958058,Nlu:48.99958058):245.4543469):44.5460725):35.82192079,(Phu:362.4323662,(Ame:345,(Tca:314.9941411,(Dme:279.4348203,Bmo:279.4348203):35.55932079):30.00585895):17.43236624):12.38955455):13.58490598,((Bge:112.528622,(Cse:61.00174814,Zne:61.00174814):51.52687381):135.471378,((Gbi:89.20882147,Lko:89.20882147):126.4915381,Lmi:215.7003595):32.29964046):140.4068268);\n",
    "\n",
    "lambda -l 0.00172492 0.000778238 0.00151124 -t  (((1,((1,1)1,(1,1)1)1)1,(2,(2,(2,(2,2)2)2)2)2)1,((3,(3,3)3)3,((3,3)3,3)3)3);\n",
    "\n",
    "report reports/log_V2.2.4_largefamilies\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "report reports/log_V2.2.4_largefamilies\n",
    "\n",
    "\n",
    "```\n",
    "\n",
    "\n",
    "Run:\n",
    "\n",
    "```\n",
    "~/data_disk/Software/CAFE/release/cafe CAFE_v2.2.4_largefamilies\n",
    "\n",
    "```\n",
    "\n",
    "### RESULTS for Lareg Families\n",
    "\n",
    "Process results \n",
    "\n",
    "```\n",
    "python  cafetutorial_report_analysis.py -i reports/log_V2.2.4_largefamilies.cafe -o reports/summary_V2.2.4_largefamilies\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Draw tree\n",
    "\n",
    "usedtree=\"(((Foc:339,((Api:40.38991217,Sfl:40.38991217):254.0640153,(Lst:48.99958058,Nlu:48.99958058):245.4543469):44.5460725):35.82192079,(Phu:362.4323662,(Ame:345,(Tca:314.9941411,(Dme:279.4348203,Bmo:279.4348203):35.55932079):30.00585895):17.43236624):12.38955455):13.58490598,((Bge:112.528622,(Cse:61.00174814,Zne:61.00174814):51.52687381):135.471378,((Gbi:89.20882147,Lko:89.20882147):126.4915381,Lmi:215.7003595):32.29964046):140.4068268)\"\n",
    "cafenodetree='(((Foc<0>,((Api<2>,Sfl<4>)<3>,(Lst<6>,Nlu<8>)<7>)<5>)<1>,(Phu<10>,(Ame<12>,(Tca<14>,(Dme<16>,Bmo<18>)<17>)<15>)<13>)<11>)<9>,((Bge<20>,(Cse<22>,Zne<24>)<23>)<21>,((Gbi<26>,Lko<28>)<27>,Lmi<30>)<29>)<25>)<19>'\n",
    "\n",
    "# rapid evolving\n",
    "%run cafetutorial_draw_tree.py -i reports/summary_V2.2.4_largefamilies_node.txt -t $usedtree -d $cafenodetree -o reports/summary_V2.2.4_largefamilies_Rapid.png -y Rapid\n",
    "# expansions\n",
    "%run cafetutorial_draw_tree.py -i reports/summary_V2.2.4_largefamilies_node.txt -t $usedtree -d $cafenodetree -o reports/summary_V2.2.4_largefamilies_Expanisons.png -y Expansions\n",
    "# #Contractions\n",
    "%run cafetutorial_draw_tree.py -i reports/summary_V2.2.4_largefamilies_node.txt -t $usedtree -d $cafenodetree -o reports/summary_V2.2.4_largefamilies_Contractions.png -y Contractions\n",
    "\n",
    "## Ignore warnings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## PARSE RESULTS MYSELF\n",
    "\n",
    "### I can get the OGS significantly gained/lost at any node\n",
    "\n",
    "#Node=\"Lko\"\n",
    "Node=\"<27>\"# Gbi+Lko\n",
    "#Node=\"<29>\"# Gbi+Lko+Lmi\n",
    "\n",
    "with open(\"reports/summary_V2.2.4_largefamilies_fams.txt\") as f:\n",
    "    for line in f:\n",
    "        #print(line)\n",
    "        #f line.startswith(\"<\"):\n",
    "            #print(line)\n",
    "        if line.startswith(Node):\n",
    "            print(line)\n",
    "            Expan=0\n",
    "            Reduc=0\n",
    "            OGs=line.split(\",\")\n",
    "            for items in OGs:\n",
    "                itemsplit=items.split(\"[\")[1].split(\"*\")[0]\n",
    "                print(itemsplit)\n",
    "                if \"+\" in itemsplit:\n",
    "                    Expan+= 1\n",
    "                if \"-\" in itemsplit:\n",
    "                    Reduc+= 1\n",
    "            print(\"Number of OGs=\", len(OGs))\n",
    "            print(\"Number of OG Expantions= \",Expan, \" Number of OG Contractions= \", Reduc )\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Print All the OGs significant in each branch  in files\n",
    "\n",
    "\n",
    "#CAFE_resultsfile=\"reports/summary_V2.2.4_largefamilies_fams.txt\"\n",
    "#CAFE_resultsfile=\"reports/summary_run2_aftererror_fams.txt\"\n",
    "\n",
    "\n",
    "linenum=0\n",
    "with open(CAFE_resultsfile) as f:\n",
    "    for line in f:\n",
    "        linenum=+1\n",
    "        line=line.rstrip()\n",
    "        #print(line)\n",
    "        #f line.startswith(\"<\"):\n",
    "            #print(line)\n",
    "        if \"<\" in line.split(\"\\t\")[0] :\n",
    "            node=line.split(\":\")[0]\n",
    "            print(\"Node= \",node)\n",
    "            #filename_expa=\"Expansions_Contractions/Expansion_node\"+node+\".txt\"\n",
    "            #filename_cont=\"Expansions_Contractions/Contraction_node\"+node+\".txt\"\n",
    "            #filename_expa=\"Expansions_Contractions/Largefamily_Expansion_node\"+node+\".txt\"\n",
    "            #filename_cont=\"Expansions_Contractions/Largefamily_Contraction_node\"+node+\".txt\"\n",
    "            f1= open(filename_expa,\"w+\")\n",
    "            f2= open(filename_cont,\"w+\")\n",
    "        else:\n",
    "            node=\"no node line\"\n",
    "        try:\n",
    "            #print(line)q\n",
    "            Expan=0\n",
    "            Reduc=0\n",
    "            OGs=line.split(\",\")\n",
    "\n",
    "            for items in OGs:\n",
    "                itemsplit=items.split(\"[\")[1].split(\"*\")[0]\n",
    "                if \"+\" in itemsplit:\n",
    "                    Expan+= 1\n",
    "                    print(itemsplit,\"-->\", items.split(\"[\")[0], \" -> Expansion\")\n",
    "                    f1.write(items.split(\"[\")[0]+\"\\t\"+itemsplit+\"\\n\")\n",
    "                if \"-\" in itemsplit:\n",
    "                    Reduc+= 1\n",
    "                    print(itemsplit,\"-->\",  items.split(\"[\")[0] , \" -> Contraction\")\n",
    "                    f2.write(items.split(\"[\")[0]+\"\\t\"+itemsplit+\"\\n\")\n",
    "            f1.close() \n",
    "            f2.close() \n",
    "            print(\"Number of OGs=\", len(OGs))\n",
    "            print(\"Number of OG Expantions= \",Expan, \" Number of OG Contractions= \", Reduc )\n",
    "        except:\n",
    "            print(\"\\n Line\", linenum, \"does not contain a branch:\")\n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "filename_cont=\"Expansions_Contractions/Contraction_node\"+node+\".txt\"    \n",
    "print(\"\".join(filename_cont) )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Functional Analysis Expansions and Contractions\n",
    "\n",
    "After identifying gene ffamily expansions and contractions, I would like to know their functions.\n",
    "\n",
    "\n",
    "I installed interproscan-5.36-75\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## FUNCTIONS later called to annotate OGs\n",
    "\n",
    "####################\n",
    "# Function to process Interproscan output\n",
    "def fucntionSummarizeIpr(OGipr):\n",
    "    outputdict=dict()\n",
    "    pfamDesc=dict()\n",
    "    pfamID=dict()\n",
    "    godict=dict()\n",
    "    godictnames=dict()\n",
    "    seqiddict=dict()\n",
    "    gos=[]\n",
    "\n",
    "    #OGipr=\"OG0000026.fa.tsv\"\n",
    "    file=\"Expansions_Contractions/Interproscan5_output/%s\" %OGipr \n",
    "\n",
    "    with open(file)as f:\n",
    "        for lines in f:\n",
    "            lines=lines.rstrip()\n",
    "            fields=lines.split(\"\\t\")\n",
    "            seqid=fields[0]\n",
    "            #print(seqid)\n",
    "            if seqid not in seqiddict:\n",
    "                seqiddict[seqid]=1\n",
    "                lines=lines.rstrip()\n",
    "                Org=fields[0].split(\"_\")[0]\n",
    "                pfam=fields[4]\n",
    "                if pfam in pfamID:\n",
    "                    pfamID[pfam]=pfamID[pfam]+1\n",
    "                else:\n",
    "                    pfamID[pfam]=1\n",
    "                    pfamDesc[pfam]=fields[5]\n",
    "\n",
    "                try:\n",
    "                    gos=fields[13].split(\"|\")\n",
    "                    for got in gos:\n",
    "                        if got in godict:\n",
    "                            godict[got]=godict[got]+1\n",
    "                        else:\n",
    "                            godict[got]=1  \n",
    "\n",
    "                except:\n",
    "                    gos=\"NA\"\n",
    "                    #print(\"No GO\", lines)\n",
    "                #print(Org, seqid,pfam, pfamd, gos)\n",
    "    print(\"Number of seqsin OG:\", len(seqiddict))\n",
    "\n",
    "    for key in pfamID:\n",
    "        print(key , pfamID[key],pfamDesc[key])\n",
    "    for key in godict:\n",
    "        print(key , godict[key],getGOinfo(key)[1], getGOinfo(key)[2])\n",
    "    for key in godict:\n",
    "        godictnames[key]= [ godict[key], getGOinfo(key)[1], getGOinfo(key)[2]]\n",
    "        \n",
    "    outputdict[\"pfamids\"]=pfamID \n",
    "    outputdict[\"pfamdesc\"]=pfamDesc \n",
    "    outputdict[\"goterms\"]=godictnames \n",
    "\n",
    "    return(outputdict)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "#ipprs_cancommand=\"~/data_disk/Software/my_interproscan/interproscan-5.36-75.0/interproscan.sh -d Interproscan5_output --goterms -appl Pfam -i  Contraction_nodeGbi\\<6\\>.txt\" % INPUTFILE\n",
    "\n",
    "\n",
    "############# function run interproscan and get output\n",
    "def iprscanfunc(file):\n",
    "    OGseqsfile=OGseqsfileDir+file\n",
    "    copyfile(OGseqsfile, \"Expansions_Contractions/OG_seqs/\"+file)\n",
    "    replace_stars_command='sed  -i \"s/\\*//g\" %s' % \"Expansions_Contractions/OG_seqs/\"+file\n",
    "    print(replace_stars_command)\n",
    "    subprocess.run(replace_stars_command,shell=True)\n",
    "    ipprs_cancommand=\"~/data_disk/Software/my_interproscan/interproscan-5.36-75.0/interproscan.sh  -f TSV -d Expansions_Contractions/Interproscan5_output --goterms -appl Pfam -i  %s\" % \"Expansions_Contractions/OG_seqs/\"+file\n",
    "    print(ipprs_cancommand)\n",
    "    subprocess.run(ipprs_cancommand,shell=True)\n",
    "    \n",
    "####################\n",
    "## Function to get GO names and descriptors from GO term\n",
    "def getGOinfo(GOterm):\n",
    "    responseBody=[]\n",
    "    requestURL = \"https://www.ebi.ac.uk/QuickGO/services/ontology/go/terms/%s\" % GOterm\n",
    "    r = requests.get(requestURL, headers={ \"Accept\" : \"application/json\"})\n",
    "\n",
    "    if not r.ok:\n",
    "      r.raise_for_status()\n",
    "      sys.exit()\n",
    "\n",
    "    responseBody = r.json()\n",
    "    \n",
    "    return(responseBody[\"results\"][0][\"id\"], responseBody[\"results\"][0][\"name\"], responseBody[\"results\"][0][\"definition\"][\"text\"] )\n",
    "\n",
    "####################\n",
    "\n",
    "\n",
    "## I downlaod Dmel Id mapping file  wget ftp://ftp.flybase.net/releases/current/precomputed_files/genes/fbgn_annotation_ID_fb_2019_03.tsv.gz\n",
    "\n",
    "# Function for given a drosophila gene ID (GCXXX) get the FlyBase id (using ID file) and getting summary using API\n",
    "\n",
    "#file=\"OG0000001.fa\"\n",
    "def checkflybase(file):\n",
    "\n",
    "    data = pd.read_csv(\"fbgn_annotation_ID_fb_2019_03.tsv\", sep=\"\\t\",skiprows=4, engine='python') \n",
    "    data.head()\n",
    "\n",
    "    FlybaseDictOut= dict()\n",
    "\n",
    "\n",
    "    OGseqsfile=OGseqsfileDir+file\n",
    "    print(OGseqsfile)\n",
    "    with open(OGseqsfile, \"r\") as fp:\n",
    "        for line in fp:\n",
    "            if line.startswith(\">Dme\"):\n",
    "                GCid=line.split(\"_\")[1].strip()\n",
    "                FlybaseGB=data[data['annotation_ID'] == GCid ]['primary_FBgn#'].item()\n",
    "                GeneSymbol=data[data['annotation_ID'] == GCid ]['##gene_symbol'].item()\n",
    "                print(GCid,FlybaseGB)\n",
    "                #using flybase API to get gene summary\n",
    "                requestURL = \"https://api.flybase.org/api/v1.0/gene/summaries/auto/%s\" % FlybaseGB\n",
    "                try:\n",
    "                    responseBody = requests.get(requestURL, headers={ \"Accept\" : \"application/json\"}).json()\n",
    "                    GeneSummary=responseBody[\"resultset\"][\"result\"][0]['summary']\n",
    "                except:\n",
    "                    GeneSummary=\"NA\"\n",
    "\n",
    "                tempdict=dict()\n",
    "                tempdict={GCid: {\"FlybaseGB\":FlybaseGB, \"GeneSymbol\":GeneSymbol, \"GeneSummary\":GeneSummary}}\n",
    "                FlybaseDictOut.update(tempdict)\n",
    "            \n",
    "    return(FlybaseDictOut)\n",
    "####################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import re, os, subprocess, sys,glob\n",
    "from shutil import copyfile\n",
    "import requests, sys, json\n",
    "\n",
    "\n",
    "OGseqsfileDir='/Orthofinder/Orthofinder_input_output/V2.2/OrthoFinder/Results_Jul30/Orthogroup_Sequences/'\n",
    "\n",
    "CAFE_resultsfile=\"reports/summary_V2.2.4_corrected_fams.txt\"\n",
    "\n",
    "#CAFE_resultsfile=\"reports/summary_V2.2.4_largefamilies_fams.txt\"\n",
    "\n",
    "\n",
    "linenum=0\n",
    "with open(CAFE_resultsfile) as f:\n",
    "    for line in f:\n",
    "        done=False\n",
    "        linenum=linenum+1\n",
    "        line=line.rstrip()\n",
    "        #print(line)\n",
    "        #f line.startswith(\"<\"):\n",
    "            #print(line)\n",
    "        if \"<\" in line.split(\"\\t\")[0] :\n",
    "            node=line.split(\":\")[0]\n",
    "            print(\"\\n\\n Node= \",node, \"\\n\\n\")\n",
    "#             filename_expa=\"Expansions_Contractions/Largefamily_Expansion_node\"+node+\".txt\"\n",
    "#             filename_cont=\"Expansions_Contractions/Largefamily_Contraction_node\"+node+\".txt\"\n",
    "            filename_expa=\"Expansions_Contractions/Expansion_node\"+node+\".txt\"\n",
    "            filename_cont=\"Expansions_Contractions/Contraction_node\"+node+\".txt\"\n",
    "            #### Check if node alreay analyzed (to avou repeating if fails)\n",
    "            if os.path.exists(filename_expa):\n",
    "                print(\"Previously computed!\")\n",
    "                done=True \n",
    "                \n",
    "            else:    \n",
    "                f1= open(filename_expa,\"w+\")\n",
    "                f2= open(filename_cont,\"w+\")\n",
    "                f1.write(\"Number of expanded OG : \"+str(line.count(\"+\"))+\"\\n\")\n",
    "                f2.write(\"Number of contracted OG : \"+str(line.count(\"-\") )+\"\\n\")\n",
    "        else:\n",
    "            node=\"no node line\"\n",
    "        if \">\" in line and not \"#\" in line and done == False:\n",
    "            print(line)\n",
    "            Expan=0\n",
    "            Reduc=0\n",
    "            OGs=line.split(\",\")\n",
    "\n",
    "            for items in OGs:\n",
    "                try: \n",
    "                    itemsplit=items.split(\"[\")[1].split(\"*\")[0]\n",
    "                except:\n",
    "                    print(\"No expantions/contractions\")\n",
    "                    itemsplit=\"No expansions or contractions\"\n",
    "                if \"+\" in itemsplit:\n",
    "                    Expan+= 1\n",
    "                    print(itemsplit,\"-->\", items.split(\"[\")[0], \" -> Expansion\")\n",
    "                    ### Fucntional Annotation interporscan   \n",
    "                    fastafilename= re.sub(r'.*OG', 'OG', items)#.split(\"[\")[0]+\".fa\"\n",
    "                    fastafilename=fastafilename.split(\"[\")[0]+\".fa\"\n",
    "                    f1.write(fastafilename+\"\\n\")\n",
    "                    f1.write(\"\\tExpansion size: \"+itemsplit+\"\\n\")\n",
    "                    \n",
    "                    print(fastafilename)\n",
    "                    iprscanfunc(fastafilename)\n",
    "                    IPRresults= fucntionSummarizeIpr(fastafilename+\".tsv\")\n",
    "                    if len(IPRresults['pfamids'].keys())>0:                   \n",
    "                        f1.write(\"\\t\\t Num. of Genes with the PFAM \\t PFAM ID \\t PFAM descriptor\\n\")\n",
    "                        for key in IPRresults['pfamids']:\n",
    "                            f1.write(\"\\t\\t\"+str(IPRresults['pfamids'][key])+\"\\t\"+key+\"\\t\"+str(IPRresults['pfamdesc'][key])+\"\\n\")\n",
    "                    if len(IPRresults['goterms'].keys())>0:\n",
    "                        f1.write(\"\\t\\t Num. of Genes with the GOterm \\t GOterms \\t GOterm name \\t GO Term descriptor\\n\")\n",
    "                        for key in IPRresults['goterms']:\n",
    "                            f1.write(\"\\t\\t\"+str(IPRresults['goterms'][key][0])+\"\\t\"+key+\"\\t\"+IPRresults['goterms'][key][1]+\"\\t\"+IPRresults['goterms'][key][2]+\"\\n\")                   \n",
    "                    FlyBaseInfo=checkflybase(fastafilename)# check if any gene in OG is from dmel and get info from flybase\n",
    "                    if len(FlyBaseInfo.keys())>0:                   \n",
    "                        f1.write(\"\\t\\t Dmel Genes in OG \\t Gene ID \\t Flybase ID \\t GeneSymbol \\t Flybase Summary\\n\")\n",
    "                        for key in FlyBaseInfo:\n",
    "                            f1.write(\"\\t\\t\\t\"+key+\"\\t\"+str(FlyBaseInfo[key]['FlybaseGB'])+\"\\t\"+str(FlyBaseInfo[key]['GeneSymbol'])+\"\\t\"+str(FlyBaseInfo[key]['GeneSummary'])+\"\\n\" )            \n",
    "                 \n",
    "                if \"-\" in itemsplit:\n",
    "                    Reduc+= 1\n",
    "                    print(itemsplit,\"-->\",  items.split(\"[\")[0] , \" -> Contraction\")\n",
    "\n",
    "\n",
    "                    ### Fucntional Annotation interporscan   \n",
    "                    fastafilename= re.sub(r'.*OG', 'OG', items)#.split(\"[\")[0]+\".fa\"\n",
    "                    fastafilename=fastafilename.split(\"[\")[0]+\".fa\"\n",
    "                    f2.write(fastafilename+\"\\n\")\n",
    "                    f2.write(\"\\tContraction size: \"+itemsplit+\"\\n\")\n",
    "                    \n",
    "                    print(fastafilename)\n",
    "                    iprscanfunc(fastafilename)\n",
    "                    IPRresults= fucntionSummarizeIpr(fastafilename+\".tsv\")\n",
    "                    if len(IPRresults['pfamids'].keys())>0:                   \n",
    "                        f2.write(\"\\t\\t Num. of Genes with the PFAM \\t PFAM ID \\t PFAM descriptor\\n\")\n",
    "                        for key in IPRresults['pfamids']:\n",
    "                            f2.write(\"\\t\\t\"+str(IPRresults['pfamids'][key])+\"\\t\"+key+\"\\t\"+str(IPRresults['pfamdesc'][key])+\"\\n\")\n",
    "                    if len(IPRresults['goterms'].keys())>0:                  \n",
    "                        f2.write(\"\\t\\t Num. of Genes with the GOterm \\t GOterms \\t GOterm name \\t GO Term descriptor\\n\")\n",
    "                        for key in IPRresults['goterms']:\n",
    "                            f2.write(\"\\t\\t\"+str(IPRresults['goterms'][key][0])+\"\\t\"+key+\"\\t\"+IPRresults['goterms'][key][1]+\"\\t\"+IPRresults['goterms'][key][2]+\"\\n\")                   \n",
    "                    FlyBaseInfo=checkflybase(fastafilename)# check if any gene in OG is from dmel and get info from flybase\n",
    "                    if len(FlyBaseInfo.keys())>0:\n",
    "                        f2.write(\"\\t\\t Dmel Genes in OG \\t Gene ID \\t Flybase ID \\t GeneSymbol \\t Flybase Summary\\n\")\n",
    "                        for key in FlyBaseInfo:\n",
    "                            f2.write(\"\\t\\t\\t\"+key+\"\\t\"+str(FlyBaseInfo[key]['FlybaseGB'])+\"\\t\"+str(FlyBaseInfo[key]['GeneSymbol'])+\"\\t\"+str(FlyBaseInfo[key]['GeneSummary'])+\"\\n\" )            \n",
    "\n",
    "                    #f2.write(items.split(\"[\")[0]+\"\\t\"+itemsplit+\"\\n\")\n",
    "            f1.close() \n",
    "            f2.close() \n",
    "            print(\"Number of OGs=\", len(OGs))\n",
    "            print(\"Number of OG Expantions= \",Expan, \" Number of OG Contractions= \", Reduc )\n",
    "        else:\n",
    "            print(\"\\n Line\", linenum, \"does not contain a branch:\")\n",
    "            \n",
    "\n",
    "        \n",
    "#ipprs_cancommand=\"~/data_disk/Software/my_interproscan/interproscan-5.36-75.0/interproscan.sh -d Interproscan5_output --goterms -appl Pfam -i  Contraction_nodeGbi\\<6\\>.txt\" % INPUTFILE\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# References \n",
    "\n",
    "CAFE Tutorial: https://iu.app.box.com/v/cafetutorial-files\n",
    "\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
