---
title: "CpGoe Multiple species"
author: "Guillem Ylla"
date: "11/26/2019"
output:
  html_document:
    df_print: paged
    toc: true
    toc_depth: 6
  pdf_document: default
  word_document: default
---


```{r loadlibs, include=FALSE, cache=FALSE, message=F, warning=F}
setwd("/home/guillem/CpGoe_Foc_forMSreview/")
library(RSQLite)
library(Biostrings)
library(ggplot2)
library(grid)
library(gridExtra)
library(ggwordcloud)
library(RColorBrewer)
library(tagcloud)
library(mixtools)
library(data.table)
library(dplyr)
#library(Rgraphviz)
library(knitr)
library(reconPlots)
library(topGO)
library(forcats)
```
```{r setup, include=FALSE, cache=FALSE, message=F, warning=F}
knitr::opts_chunk$set(cache = TRUE)
knitr::opts_chunk$set(echo = TRUE)# by deafult echo=true in all chunks
opts_chunk$set(tidy.opts=list(width.cutoff=60),tidy=TRUE)# avoid code outside page
```
.


## Load CpGoe values

CpGoe calculated in: [Count_CpGoe.ipynb](https://github.com/guillemylla/Crickets_Genome_Annotation/blob/master/Analysis/CpGoe/1_Count_CpGoe.ipynb)


```{r, GbiCpG, include=TRUE, cache=TRUE, warning=FALSE, message = FALSE}


CpGoe_insects<-read.delim("CpGoe_outputs/All_insects_CpGpe")
CpGoe_insects<-CpGoe_insects[!is.na(CpGoe_insects$CpGoe),]
```

## Overview CpGoe 16 Insects


Plot the CpG distribution for all 16 insects:

```{r, CpG,  include=TRUE, cache=TRUE, warning=FALSE, message = FALSE}
CGoe_plot1<-ggplot(CpGoe_insects, aes(x=CpGoe,  y = ..density.. )) + 
   facet_wrap(~Spp)+
   geom_histogram(bins=1000)+
   scale_x_continuous(name = "CpGo/e") +
   ggtitle("CGo/e distribution")  +ylab("Number of CDS")
CGoe_plot1
```


Same, but removing genes CpGoe>2.


```{r, CpG2,  include=TRUE, cache=TRUE, warning=FALSE, message = FALSE}
dim(CpGoe_insects[CpGoe_insects$CpGoe>2,])

CpGoe_insects_nolarger2<-CpGoe_insects[CpGoe_insects$CpGoe<2,]

CpGoe_insects_b02<-CpGoe_insects[CpGoe_insects$CpGoe<2 & CpGoe_insects$CpGoe>0,]


CGoe_plot2<-ggplot(CpGoe_insects_b02, aes(x=CpGoe,  y = ..density.. )) + 
   facet_wrap(~Spp)+
   geom_histogram(bins=150)+
   scale_x_continuous(name = "CpGo/e") +
   ggtitle("CGo/e distribution")  +ylab("Density")
CGoe_plot2
```

```{r, CpG_Miniature_forpaper,  include=TRUE, cache=TRUE, warning=FALSE, message = FALSE}

CGoe_plotMiniatures<-ggplot(CpGoe_insects_b02, aes(x=CpGoe,  y = ..density.. )) + 
   facet_wrap(~Spp)+
   geom_histogram(bins=150, colour="black")+
   scale_x_continuous(name = "CpGo/e") +
  ggtitle("CGo/e distribution")  +ylab("Density")+
theme(
    #plot.background = element_rect(fill = "transparent", color = NA), # bg of the plot
    panel.grid.major = element_blank(), # get rid of major grid
    panel.grid.minor = element_blank(), # get rid of minor grid
    panel.background = element_rect(fill = "transparent",colour = NA),
    plot.background = element_rect(fill = "transparent",colour = NA),
    axis.text.x=element_blank(),
    axis.text.y=element_blank(),
    axis.ticks.x=element_blank(),
   axis.ticks.y=element_blank(),
    legend.background = element_rect(fill = "transparent"), # get rid of legend bg
    legend.box.background = element_rect(fill = "transparent") # get rid of legend panel bg
  )

```


## CpGoe bimodal model 16 insects (Fig 2A)


Lets try to fit all of them in a bimodal distribution with a gaussian Mixture model. 


```{r,bimodal,  message=FALSE, warning=FALSE,  eval=FALSE, cache=TRUE, include=TRUE, paged.print=FALSE}

for (species in unique(CpGoe_insects$Spp)) {
  print(species)
  # only cpg values >0 & <2
  CpGoe_Spp<-CpGoe_insects_b02[CpGoe_insects_b02$Spp==species,] 
  #CpGoe_Spp<-CpGoe_insects[CpGoe_insects$Spp==species,]
  CpGoe_toplot = CpGoe_Spp$CpGoe
  CpGoe_toplot<-CpGoe_toplot[!is.na(CpGoe_toplot)]
  mixmdl= normalmixEM(CpGoe_toplot,maxit=20000)

  #Standard Error function from mixtools bootsratping 100 
  se<-boot.se(mixmdl,B=100)
  # Intervals
  CI_1_right=mixmdl$mu[1]+(se$mu.se[1]*2.8)
  CI_1_left=mixmdl$mu[1]-(se$mu.se[1]*2.8)
  
  CI_2_right=mixmdl$mu[2]+(se$mu.se[2]*2.8)
  CI_2_left=mixmdl$mu[2]-(se$mu.se[2]*2.8)


  ## function for ggplots:
  plot_mix_comps <- function(x, mu, sigma, lam) {
      lam * dnorm(x, mu, sigma)}

  
   Biomdal_dist<-data.frame(x = mixmdl$x) %>%
              ggplot() +
              geom_histogram(aes(x, ..density..),  colour="black", alpha=.8, bins=120) +
              stat_function(geom = "line", fun = plot_mix_comps,
                            args = list(mixmdl$mu[1], mixmdl$sigma[1], lam = mixmdl$lambda[1]),
                            colour = "#E69F00", lwd = 1) +
              stat_function(geom = "line", fun = plot_mix_comps,
                            args = list(mixmdl$mu[2], mixmdl$sigma[2], lam = mixmdl$lambda[2]),
                            colour = "#56B4E9", lwd = 1) +
              geom_vline(xintercept = mixmdl$mu[1], col = "#E69F00", size = .8) + 
              annotate("text",label=round(mixmdl$mu[1], 3),y=  1.60, x=mixmdl$mu[1] + .01, col="#E69F00",size=4)+
              annotate("rect", xmin = CI_1_left, xmax =CI_1_right, ymin = 0, ymax = Inf,   alpha = 0.2, fill = "red")    +
              geom_vline(xintercept = mixmdl$mu[2], col = "#56B4E9", size = .8)+
              annotate("rect", xmin = CI_2_left, xmax = CI_2_right, ymin = 0, ymax = Inf,   alpha = 0.2, fill = "red")    +
              annotate("text",label=round(mixmdl$mu[2], 3),y=  1.20, x=mixmdl$mu[2] + .01, col="#56B4E9",size=4)+
              annotate("text",label=round(round(mixmdl$lambda[1], 2), 3),y=  1.60, x=mixmdl$mu[1] - .3, col="black",size=4)+
              annotate("text",label=round(round(mixmdl$lambda[2], 2), 3),y=  1.20, x=mixmdl$mu[2] + .3, col="black",size=4)+
             ylab("Density")+xlab("CpGo/e")+ggtitle(species)


#red line spearateing
data<-ggplot_build(Biomdal_dist)$data[[1]]
data_int<-data[data$x>mixmdl$mu[1] & data$x<mixmdl$mu[2] ,]
cpg_threshold<-data_int[data_int$y==min(data_int$y),]$x
Biomdal_dist<-Biomdal_dist+ geom_vline(xintercept = cpg_threshold, col = "red", size = .5) +
                          annotate("text",label=round(cpg_threshold, 3),y=  1.3, x=cpg_threshold + .01, col="red",size=4)


#ggsave(paste0("plots/",paste0(species, "_CpGplot.png")), Biomdal_dist,  width = 9, height =5, units ="in",dpi = 300,)

}
```

The Gussian Mixture model, fits 2 gaussian ditributions to the CpGoe of each insect.

I get the mean of each curve (vertical bars) an calculate the standard error (SE) of each mean  with 100 bootstraps  The SE is used to calculate the 99% confidence interval around the mean (red shade).

The red line represents the intersection between the 2 curves, and it is used as threshold between genes with **Low and High CpGoe**.


```{r, addplots, results = "asis",  cache=FALSE}

plots <- list.files("CpGoe/plots/")

for(i in plots){
  if(grepl("_CpGplot.png", i)){
      filename <- file.path("CpGoe/plots", i)
      cat("![text](",filename,")\n")
  }
}

```


## Gryllus bimaculatus CpGoe distribution


```{r, gbibimodalgbi,  include=TRUE, cache=TRUE, warning=FALSE,  results='hide', message = FALSE}

Gbi_CpG<-CpGoe_insects_b02[CpGoe_insects_b02$Spp=="Gryllus_bimaculatus",]

CpGoe_toplotGbi = Gbi_CpG$CpGoe
CpGoe_toplotGbi<-CpGoe_toplotGbi[!is.na(CpGoe_toplotGbi)]
mixmdlGbi = normalmixEM(CpGoe_toplotGbi,maxit=20000)


 ## Calculating SE using mixtools bootystrap function
se<-boot.se(mixmdlGbi,B=100)
#confidence intervals
CI_1_right=mixmdlGbi$mu[1]+(se$mu.se[1]*2.8)
CI_1_left=mixmdlGbi$mu[1]-(se$mu.se[1]*2.8)

CI_2_right=mixmdlGbi$mu[2]+(se$mu.se[2]*2.8)
CI_2_left=mixmdlGbi$mu[2]-(se$mu.se[2]*2.8)


## In ggplots:
plot_mix_comps <- function(x, mu, sigma, lam) {
  lam * dnorm(x, mu, sigma)
}


Bimodal_Gbi<-data.frame(x = mixmdlGbi$x) %>%
              ggplot() +
              geom_histogram(aes(x, ..density..),  colour="black", alpha=.8, bins=100) +
              stat_function(geom = "line", fun = plot_mix_comps,
                            args = list(mixmdlGbi$mu[1], mixmdlGbi$sigma[1], lam = mixmdlGbi$lambda[1]),
                            colour = "#E69F00", lwd = 1) +
              stat_function(geom = "line", fun = plot_mix_comps,
                            args = list(mixmdlGbi$mu[2], mixmdlGbi$sigma[2], lam = mixmdlGbi$lambda[2]),
                            colour = "#56B4E9", lwd = 1) +
              geom_vline(xintercept = mixmdlGbi$mu[1], col = "#E69F00", size = .8) + 
              annotate("rect", xmin = CI_1_left, xmax = CI_1_right, ymin = 0, ymax = Inf,   alpha = 0.2, fill = "red")    +
              annotate("text",label=round(mixmdlGbi$mu[1], 3),y=  1.8, x=mixmdlGbi$mu[1] + .1, col="#E69F00",size=4)+
              geom_vline(xintercept = mixmdlGbi$mu[2], col = "#56B4E9", size = .8)+
              annotate("rect", xmin = CI_2_left, xmax = CI_2_right, ymin = 0, ymax = Inf,   alpha = 0.2, fill = "red")    +
              annotate("text",label=round(mixmdlGbi$mu[2], 3),y= 1.8, x=mixmdlGbi$mu[2] + .1, col="#56B4E9",size=4)+
              ylab("Density")+xlab("CpGo/e")+ylim(0,2.2)


### I separate the 2 distributions by the lowest density between means
curve1 <- function(x)  mixmdlGbi$lambda[1] * dnorm(x, mixmdlGbi$mu[1],  mixmdlGbi$sigma[1])
curve2 <- function(x)  mixmdlGbi$lambda[2] * dnorm(x, mixmdlGbi$mu[2],  mixmdlGbi$sigma[2])

curveintersectGbi<-curve_intersect(curve1, curve2, empirical = FALSE, domain = c(0, 5))

Gbi_cpg_threshold=curveintersectGbi$x


Bimodal_Gbi<-Bimodal_Gbi+ geom_vline(xintercept = Gbi_cpg_threshold, col = "red", size = .5)+
                  annotate("text",label=round(Gbi_cpg_threshold, 3),y=  1.80, x=Gbi_cpg_threshold + .1, col="red",size=4)

Bimodal_Gbi


```



## Laupala kohalensis  CpGoe distribution

```{r, lkobimodallko,  include=TRUE, cache=TRUE, warning=FALSE, results='hide',  message = FALSE}
Lko_CpG<-CpGoe_insects_b02[CpGoe_insects_b02$Spp=="Laupala_kohalensis",]

CpGoe_toplotLko = Lko_CpG$CpGoe
CpGoe_toplotLko<-CpGoe_toplotLko[!is.na(CpGoe_toplotLko)]
mixmdlLko = normalmixEM(CpGoe_toplotLko,maxit=20000)

 ## Calculating SE using mixtools bootystrap function
se<-boot.se(mixmdlLko,B=100)
#confidence intervals
CI_1_right=mixmdlLko$mu[1]+(se$mu.se[1]*2.8)
CI_1_left=mixmdlLko$mu[1]-(se$mu.se[1]*2.8)

CI_2_right=mixmdlLko$mu[2]+(se$mu.se[2]*2.8)
CI_2_left=mixmdlLko$mu[2]-(se$mu.se[2]*2.8)

## In ggplots:
plot_mix_comps <- function(x, mu, sigma, lam) {
  lam * dnorm(x, mu, sigma)
}


Bimodal_Lko<-data.frame(x = mixmdlLko$x) %>%
              ggplot() +
              geom_histogram(aes(x, ..density..),  colour="black", alpha=.8, bins=100) +
              stat_function(geom = "line", fun = plot_mix_comps,
                            args = list(mixmdlLko$mu[1], mixmdlLko$sigma[1], lam = mixmdlLko$lambda[1]),
                            colour = "#E69F00", lwd = 1) +
              stat_function(geom = "line", fun = plot_mix_comps,
                            args = list(mixmdlLko$mu[2], mixmdlLko$sigma[2], lam = mixmdlLko$lambda[2]),
                            colour = "#56B4E9", lwd = 1) +
              geom_vline(xintercept = mixmdlLko$mu[1], col = "#E69F00", size = .8) + 
              annotate("rect", xmin = CI_1_left, xmax = CI_1_right, ymin = 0, ymax = Inf,   alpha = 0.2, fill = "red")    +
              annotate("text",label=round(mixmdlLko$mu[1], 3),y=  1.8, x=mixmdlLko$mu[1] + 0.1, col="#E69F00",size=4)+
              geom_vline(xintercept = mixmdlLko$mu[2], col = "#56B4E9", size = .8)+
              annotate("rect", xmin = CI_2_left, xmax = CI_2_right, ymin = 0, ymax = Inf,   alpha = 0.2, fill = "red")    +
              annotate("text",label=round(mixmdlLko$mu[2], 3),y= 1.8, x=mixmdlLko$mu[2] + 0.1 ,col="#56B4E9",size=4)+
              #annotate("text",label=paste0("LogLik=",round(mixmdlLko$loglik, 1)),y=  1.5, x=1.5, col="black",size=4)+
              ylab("Density")+xlab("CpGo/e")+ylim(0,2.2)


### I separate the 2 distributions by the lowest density between means
curve1 <- function(x)  mixmdlLko$lambda[1] * dnorm(x, mixmdlLko$mu[1],  mixmdlLko$sigma[1])
curve2 <- function(x)  mixmdlLko$lambda[2] * dnorm(x, mixmdlLko$mu[2],  mixmdlLko$sigma[2])

curveintersectLko<-curve_intersect(curve1, curve2, empirical = FALSE, domain = c(0, 5))

Lko_cpg_threshold=curveintersectLko$x


Bimodal_Lko<-Bimodal_Lko+ geom_vline(xintercept = Lko_cpg_threshold, col = "red", size = .5)+
                  annotate("text",label=round(Lko_cpg_threshold, 3),y=  1.80, x=Lko_cpg_threshold + .1, col="red",size=4)

Bimodal_Lko

```


## Apis mellifera  CpGoe distribution

```{r message=FALSE, warning=FALSE, amebimodalame,   include=TRUE, cache=TRUE, results='hide'}
Ame_CpG<-CpGoe_insects_b02[CpGoe_insects_b02$Spp=="Apis_mellifera",]

CpGoe_toplotAme = Ame_CpG$CpGoe
CpGoe_toplotAme<-CpGoe_toplotAme[!is.na(CpGoe_toplotAme)]
mixmdlAme = normalmixEM(CpGoe_toplotAme,maxit=20000)

 ## Calculating SE using mixtools bootystrap function
se<-boot.se(mixmdlAme,B=100)
#confidence intervals
CI_1_right=mixmdlAme$mu[1]+(se$mu.se[1]*2.8)
CI_1_left=mixmdlAme$mu[1]-(se$mu.se[1]*2.8)

CI_2_right=mixmdlAme$mu[2]+(se$mu.se[2]*2.8)
CI_2_left=mixmdlAme$mu[2]-(se$mu.se[2]*2.8)

## In ggplots:
plot_mix_comps <- function(x, mu, sigma, lam) {
  lam * dnorm(x, mu, sigma)
}


Bimodal_Ame<-data.frame(x = mixmdlAme$x) %>%
              ggplot() +
              geom_histogram(aes(x, ..density..),  colour="black", alpha=.8, bins=100) +
              stat_function(geom = "line", fun = plot_mix_comps,
                            args = list(mixmdlAme$mu[1], mixmdlAme$sigma[1], lam = mixmdlAme$lambda[1]),
                            colour = "#E69F00", lwd = 1) +
              stat_function(geom = "line", fun = plot_mix_comps,
                            args = list(mixmdlAme$mu[2], mixmdlAme$sigma[2], lam = mixmdlAme$lambda[2]),
                            colour = "#56B4E9", lwd = 1) +
              geom_vline(xintercept = mixmdlAme$mu[1], col = "#E69F00", size = .8) + 
              annotate("rect", xmin = CI_1_left, xmax = CI_1_right, ymin = 0, ymax = Inf,   alpha = 0.2, fill = "red")    +
              annotate("text",label=round(mixmdlAme$mu[1], 3),y=  1.8, x=mixmdlAme$mu[1] + .1, col="#E69F00",size=4)+
              geom_vline(xintercept = mixmdlAme$mu[2], col = "#56B4E9", size = .8)+
              annotate("rect", xmin = CI_2_left, xmax = CI_2_right, ymin = 0, ymax = Inf,   alpha = 0.2, fill = "red")    +
              annotate("text",label=round(mixmdlAme$mu[2], 3),y= 1.8, x=mixmdlAme$mu[2] + .1, col="#56B4E9",size=4)+
              ylab("Density")+xlab("CpGo/e")+ylim(0,2.2)


### I separate the 2 distributions by the lowest density between means

curve1 <- function(x)  mixmdlAme$lambda[1] * dnorm(x, mixmdlAme$mu[1],  mixmdlAme$sigma[1])
curve2 <- function(x)  mixmdlAme$lambda[2] * dnorm(x, mixmdlAme$mu[2],  mixmdlAme$sigma[2])

curveintersectAme<-curve_intersect(curve1, curve2, empirical = FALSE, domain = c(0, 5))

Ame_cpg_threshold=curveintersectAme$x
Bimodal_Ame<-Bimodal_Ame+ geom_vline(xintercept = Ame_cpg_threshold, col = "red", size = .5)+
                 annotate("text",label=round(Ame_cpg_threshold, 3),y=  1.8, x=Ame_cpg_threshold + .1, col="red",size=4)

Bimodal_Ame

```


## Frankliniella occidentalis

```{r, gbibimodalfoc,  include=TRUE, cache=TRUE, warning=FALSE,  results='hide', message = FALSE}

Foc_CpG<-CpGoe_insects_b02[CpGoe_insects_b02$Spp=="Frankliniella_occidentalis",]

CpGoe_toplotFoc = Foc_CpG$CpGoe
CpGoe_toplotFoc<-CpGoe_toplotFoc[!is.na(CpGoe_toplotFoc)]
mixmdlFoc = normalmixEM(CpGoe_toplotFoc,maxit=20000)


 ## Calculating SE using mixtools bootystrap function
se<-boot.se(mixmdlFoc,B=100)
#confidence intervals
CI_1_right=mixmdlFoc$mu[1]+(se$mu.se[1]*2.8)
CI_1_left=mixmdlFoc$mu[1]-(se$mu.se[1]*2.8)

CI_2_right=mixmdlFoc$mu[2]+(se$mu.se[2]*2.8)
CI_2_left=mixmdlFoc$mu[2]-(se$mu.se[2]*2.8)



## In ggplots:
plot_mix_comps <- function(x, mu, sigma, lam) {
  lam * dnorm(x, mu, sigma)
}

Bimodal_Foc<-data.frame(x = mixmdlFoc$x) %>%
              ggplot() +
              geom_histogram(aes(x, ..density..),  colour="black", alpha=.8, bins=100) +
              stat_function(geom = "line", fun = plot_mix_comps,
                            args = list(mixmdlFoc$mu[2], mixmdlFoc$sigma[1], lam = mixmdlFoc$lambda[1]),
                            colour = "#E69F00", lwd = 1) +
              stat_function(geom = "line", fun = plot_mix_comps,
                            args = list(mixmdlFoc$mu[1], mixmdlFoc$sigma[1], lam = mixmdlFoc$lambda[1]),
                            colour = "#56B4E9", lwd = 1) +
              geom_vline(xintercept = mixmdlFoc$mu[2], col = "#E69F00", size = .8) + 
              annotate("rect", xmin = CI_1_left, xmax = CI_1_right, ymin = 0, ymax = Inf,   alpha = 0.2, fill = "red")    +
              geom_vline(xintercept = mixmdlFoc$mu[1], col = "#56B4E9", size = .8)+
              annotate("rect", xmin = CI_2_left, xmax = CI_2_right, ymin = 0, ymax = Inf,   alpha = 0.2, fill = "red")    +
              ylab("Density")+xlab("CpGo/e")+ylim(0,2.2)



curve1 <- function(x)  mixmdlFoc$lambda[1] * dnorm(x, mixmdlFoc$mu[1],  mixmdlFoc$sigma[1])
curve2 <- function(x)  mixmdlFoc$lambda[2] * dnorm(x, mixmdlFoc$mu[2],  mixmdlFoc$sigma[2])

curveintersectFoc<-curve_intersect(curve1, curve2, empirical = FALSE, domain = c(0, 5))

Foc_cpg_threshold=curveintersectFoc$x


Bimodal_Foc<-Bimodal_Foc+ geom_vline(xintercept = Foc_cpg_threshold, col = "red", size = .5)
                  annotate("text",label=round(Foc_cpg_threshold, 3),y=  1.80, x=Foc_cpg_threshold + .1, col="red",size=4)

Bimodal_Foc
```


## CpGoe vs Orthology

- Assign genes no High or Low CpGoe peak based on the threshols stablished before

```{r}
Gbi_CpG <- CpGoe_insects_b02[CpGoe_insects_b02$Spp == "Gryllus_bimaculatus",     ]

Gbi_CpG$CpGoe_level <- "NA"
Gbi_CpG$CpGoe_level[Gbi_CpG$CpGoe > Gbi_cpg_threshold] <- "High"
Gbi_CpG$CpGoe_level[Gbi_CpG$CpGoe <= Gbi_cpg_threshold] <- "Low"

Gbi_CpG$Gene <- sapply(strsplit(as.character(Gbi_CpG$ID), "-"), `[`, 1)

Lko_CpG <- CpGoe_insects_b02[CpGoe_insects_b02$Spp == "Laupala_kohalensis", ]

Lko_CpG$CpGoe_level <- "NA"
Lko_CpG$CpGoe_level[Lko_CpG$CpGoe > Lko_cpg_threshold] <- "High"
Lko_CpG$CpGoe_level[Lko_CpG$CpGoe <= Lko_cpg_threshold] <- "Low"
Lko_CpG$Gene <- sapply(strsplit(as.character(Lko_CpG$ID), "-"), `[`, 1)

```


```{r, AmeHighLow}

Ame_CpG<-CpGoe_insects[CpGoe_insects$Spp=="Apis_mellifera",]

Ame_CpG$'CpGoe_level'<-"NA"
Ame_CpG$'CpGoe_level'[Ame_CpG$CpGoe>Ame_cpg_threshold ]<-"High"
Ame_CpG$'CpGoe_level'[Ame_CpG$CpGoe<=Ame_cpg_threshold ]<-"Low"

Ame_CpG$Gene<-sapply(strsplit(as.character(Ame_CpG$ID),'-'), `[`, 2)

```

```{r, FocHighLow}
Foc_CpG<-CpGoe_insects[CpGoe_insects$Spp=="Frankliniella_occidentalis",]

Foc_CpG$'CpGoe_level'<-"NA"
Foc_CpG$'CpGoe_level'[Foc_CpG$CpGoe>Foc_cpg_threshold ]<-"High"
Foc_CpG$'CpGoe_level'[Foc_CpG$CpGoe<=Foc_cpg_threshold ]<-"Low"

Foc_CpG$Gene<-Foc_CpG$ID
```


```{r, FocHighLow2}
FocAmeGbiLko_insects_CpGeo<-rbind(Foc_CpG, Ame_CpG, Lko_CpG, Gbi_CpG)

## reorder: 1st low, then high
FocAmeGbiLko_insects_CpGeo$CpGoe_level <- factor(FocAmeGbiLko_insects_CpGeo$CpGoe_level,levels = c("Low","High"))


```

- Get Orthology info

```{python, preapareOGs, eval=FALSE}
##### I need to transform OG table to to 2 column OG-gene
Gene_OG_dict = {}
with open("CpGoe_outputs/Orthogroups.txt") as f:
    for line in f:
       line=line.rstrip()
       (OG, genes) = line.split(":")
       geneeList=genes.split()
       Gene_OG_dict[OG] = geneeList

fout= open("CpGoe_outputs/Orthogroups_list_OG_gene.txt","w+")
for key in Gene_OG_dict:
  for value in Gene_OG_dict[key]:
    fout.write(key+"\t"+value+"\n")
    
fout.close()


```



- Load OGs

```{r}
OG_Gene<-read.delim("CpGoe_outputs/Orthogroups_list_OG_gene.txt", header = FALSE, stringsAsFactors = FALSE)
colnames(OG_Gene)<-c("OG","Transc")

```



- Assign Genes to their corresponding Orthogroup

```{r, AddOGtoSpGlist, results='hide'}

#rename GBI transcripts
OG_Gene$Transcript<-OG_Gene$Transc
OG_Gene[OG_Gene$Transcript %like%"Gbi",]$Transcript<-sapply(strsplit(as.character(OG_Gene[OG_Gene$Transc %like%"Gbi",'Transcript']),'Gbi_'), `[`, 2)
head(OG_Gene)
OG_Gene[OG_Gene$Transc %like%"GBI",]
#rename Lko transcripts
OG_Gene[OG_Gene$Transcript %like%"Lko",]$Transcript<-paste0("Lko_",sapply(strsplit(as.character(OG_Gene[OG_Gene$Transc %like%"Lko",'Transcript']),'Lko_Lko_'), `[`, 2))
head(OG_Gene)
OG_Gene[OG_Gene$Transc %like%"Lko",]

#rename Foc transcripts
OG_Gene[OG_Gene$Transcript %like%"Foc",]$Transcript<-sapply(strsplit(as.character(OG_Gene[OG_Gene$Transc %like%"Foc",'Transcript']),'Foc_'), `[`, 2)
head(OG_Gene)
OG_Gene[OG_Gene$Transc %like%"Foc",]

# rename Ame transcipts
Ame_cds_gene_dic<-read.delim("Ame_Gene_CDS_dict.txt", header = FALSE, sep=" ")
head(Ame_cds_gene_dic)
Ame_cds_gene_dic$LOC<-paste0("Ame_",Ame_cds_gene_dic$V1)

FocAmeGbiLko_insects_CpGeo_amerenames<-merge(FocAmeGbiLko_insects_CpGeo, Ame_cds_gene_dic, by.x="Gene", by.y="V2")
head(FocAmeGbiLko_insects_CpGeo_amerenames)

OG_Gene[OG_Gene$Transc%like%"Ame",]

Ame_DF<-FocAmeGbiLko_insects_CpGeo_amerenames[,c("Spp", "Abb","LOC","Len","Gfreq","Cfreq","CGfreq","CpGoe","CpGoe_level","ID")]
colnames(Ame_DF)<-colnames(FocAmeGbiLko_insects_CpGeo)
#remove ame
dim(FocAmeGbiLko_insects_CpGeo)
FocAmeGbiLko_insects_CpGeo_v2<-FocAmeGbiLko_insects_CpGeo[!FocAmeGbiLko_insects_CpGeo$Abb=="Ame",]
dim(FocAmeGbiLko_insects_CpGeo_v2)
# Add Ame with new names
FocAmeGbiLko_insects_CpGeo_v3<-rbind(FocAmeGbiLko_insects_CpGeo_v2, Ame_DF)
dim(FocAmeGbiLko_insects_CpGeo_v3)


AmeGbiLko_CpGeo_OG<-merge(FocAmeGbiLko_insects_CpGeo_v3, OG_Gene, by.x="ID", by.y="Transcript")

table(AmeGbiLko_CpGeo_OG$Spp)
dim(Gbi_CpG) #all matched woth gene and OG
dim(Lko_CpG) #all matched woth gene and OG
dim(Ame_CpG) # I lost just 1 gene
dim(Foc_CpG) #all matched woth gene and OG
```

### CpGoe and proportion of species-specific genes (Fig 2C)

- The Orthogroups.txt generated by OrthoFinder2 assigns ALL genes to OG. Genes that have no ortholog or paralog have their own OG.
 
```{r}
NGenesInOG<-table(OG_Gene$OG)

Genes_per_OG_FocAmeGbiLko_CpGeo_OG<-AmeGbiLko_CpGeo_OG

Genes_per_OG_FocAmeGbiLko_CpGeo_OG$GenesinOG<-NGenesInOG[Genes_per_OG_FocAmeGbiLko_CpGeo_OG$OG]

```
- Get real species-specific genes

```{r}
OG_Gene_copy<-OG_Gene

OG_Gene_copy$Spp<-sapply(strsplit(OG_Gene_copy$Transc,'_'), `[`, 1)
table(OG_Gene_copy$Spp)


Spp_Specific_genes<-OG_Gene_copy %>% 
  group_by(Spp, OG) %>% 
  tally(name="Paralogs") %>% 
  group_by(OG) %>% 
  mutate(Orthologs=n()) %>% 
  filter(Orthologs==1)

dim(Spp_Specific_genes)
```



```{r}
SpSpecGenes_Met<-Genes_per_OG_FocAmeGbiLko_CpGeo_OG %>% 
  mutate(CpGoe_level=fct_relevel(CpGoe_level, c("Low","High"))) %>%
   mutate(Abb=fct_relevel(Abb, c("Gbi","Lko","Ame","Foc"))) %>%
  mutate(SpeciSpecific=ifelse(OG%in%Spp_Specific_genes$OG, "yes","no")) %>% # get species-specific
  group_by(Abb,CpGoe_level, SpeciSpecific ) %>% 
  tally(name="Num.Genes") %>% 
  mutate(freq_Speciespecific = Num.Genes / sum(Num.Genes)*100) %>% 
  filter(SpeciSpecific=="yes") %>% 
 
  ggplot(., aes(x=Abb, y=freq_Speciespecific, fill=CpGoe_level)) +
    geom_bar(stat="identity", position = "dodge")+
    ylab("% Species-Specific Genes")+#scale_fill_grey(start = 0.1,end = 0.8,)+
   scale_fill_manual(breaks = c("Low", "High"),values=c("#E69F00","#56B4E9"))+
    ggtitle("% species-specific genes per Methylation status and spp")

SpSpecGenes_Met

```

```{r}
Genes_per_OG_FocAmeGbiLko_CpGeo_OG %>% 
  mutate(CpGoe_level=fct_relevel(CpGoe_level, c("Low","High"))) %>%
   mutate(Abb=fct_relevel(Abb, c("Gbi","Lko","Ame","Foc"))) %>%
  mutate(SpeciSpecific=ifelse(OG%in%Spp_Specific_genes$OG, "yes","no")) %>% # get species-specific
  group_by(Abb,CpGoe_level, SpeciSpecific ) %>% 
  tally(name="Num.Genes") %>% 
  mutate(freq_Speciespecific = Num.Genes / sum(Num.Genes)*100) %>% 
  filter(SpeciSpecific=="yes"& CpGoe_level=="High") %>%ungroup()

Genes_per_OG_FocAmeGbiLko_CpGeo_OG %>% 
  mutate(CpGoe_level=fct_relevel(CpGoe_level, c("Low","High"))) %>%
   mutate(Abb=fct_relevel(Abb, c("Gbi","Lko","Ame","Foc"))) %>%
  mutate(SpeciSpecific=ifelse(OG%in%Spp_Specific_genes$OG, "yes","no")) %>% # get species-specific
  group_by(Abb,CpGoe_level, SpeciSpecific ) %>% 
  tally(name="Num.Genes") %>% 
  mutate(freq_Speciespecific = Num.Genes / sum(Num.Genes)*100) %>% 
  filter(SpeciSpecific=="no" & CpGoe_level=="Low" ) 
```





### UpSets plot (Fig 2B & Fig S4)


```{r}
library(UpSetR)

#write.csv(AmeGbiLko_CpGeo_OG, file="Data_for_Fig2BandS4.csv")

Data_for_Upset_v2<-list(Gbi_High=AmeGbiLko_CpGeo_OG[AmeGbiLko_CpGeo_OG$CpGoe_level=="High" & AmeGbiLko_CpGeo_OG$Abb=="Gbi", "OG"],
                     Lko_High=AmeGbiLko_CpGeo_OG[AmeGbiLko_CpGeo_OG$CpGoe_level=="High" & AmeGbiLko_CpGeo_OG$Abb=="Lko", "OG"],
                     Ame_High=AmeGbiLko_CpGeo_OG[AmeGbiLko_CpGeo_OG$CpGoe_level=="High" & AmeGbiLko_CpGeo_OG$Abb=="Ame", "OG"],
                     Foc_High=AmeGbiLko_CpGeo_OG[AmeGbiLko_CpGeo_OG$CpGoe_level=="High" & AmeGbiLko_CpGeo_OG$Abb=="Foc", "OG"],
                     Gbi_Low=AmeGbiLko_CpGeo_OG[AmeGbiLko_CpGeo_OG$CpGoe_level=="Low" & AmeGbiLko_CpGeo_OG$Abb=="Gbi", "OG"],
                     Lko_Low=AmeGbiLko_CpGeo_OG[AmeGbiLko_CpGeo_OG$CpGoe_level=="Low" & AmeGbiLko_CpGeo_OG$Abb=="Lko", "OG"],
                     Ame_Low=AmeGbiLko_CpGeo_OG[AmeGbiLko_CpGeo_OG$CpGoe_level=="Low" & AmeGbiLko_CpGeo_OG$Abb=="Ame", "OG"],
                     Foc_Low=AmeGbiLko_CpGeo_OG[AmeGbiLko_CpGeo_OG$CpGoe_level=="Low" & AmeGbiLko_CpGeo_OG$Abb=="Foc", "OG"])

Data_for_Upset_list_v2<-fromList(Data_for_Upset_v2)

Upset_raw_v3<-upset(Data_for_Upset_list_v2,   order.by = "freq", nintersects = 100,cutoff=100,
      sets = c("Ame_High","Lko_High", "Gbi_High","Foc_High","Ame_Low", "Lko_Low","Gbi_Low","Foc_Low"  ),keep.order = TRUE,
        queries = list(list(query = intersects, params = list("Gbi_Low"),  color = "#E69F00", active = T, query.name = "Low  - No horthology"), 
                  list(query = intersects, params = list("Lko_Low"),  color = "#E69F00", active = T, query.name = "Low  - No horthology") ,
                  list(query = intersects, params = list("Ame_Low"),  color = "#E69F00", active = T, query.name = "Low  - No horthology"),
                  list(query = intersects, params = list("Foc_Low"),  color = "#E69F00", active = T, query.name = "Low  - No horthology"),
                  list(query = intersects, params = list("Gbi_High"),  color = "#56B4E9", active = T, query.name = "High - No horthology"), 
                  list(query = intersects, params = list("Lko_High"),  color = "#56B4E9", active = T, query.name = "High - No horthology") ,
                  list(query = intersects, params = list("Ame_High"),  color = "#56B4E9", active = T, query.name = "High - No horthology"),
                  list(query = intersects, params = list("Foc_High"),  color = "#56B4E9", active = T, query.name = "High - No horthology"),
                  list(query = intersects, params = list("Gbi_High", "Lko_High"), color = "#56B4E9", active = T, query.name = "High CpGoe in 2"),
                  list(query = intersects, params = list("Gbi_High", "Foc_High"), color = "#56B4E9", active = T, query.name = "High CpGoe in 2"),

                  list(query = intersects, params = list("Foc_High", "Ame_High"), color = "#56B4E9", active = T, query.name = "High CpGoe in 2"),
                  list(query = intersects, params = list("Foc_High", "Gbi_High"), color = "#56B4E9", active = T, query.name = "High CpGoe in 2"),
                  list(query = intersects, params = list("Ame_High", "Gbi_High"), color = "#56B4E9", active = T, query.name = "High CpGoe in 2"),
                  list(query = intersects, params = list("Foc_High", "Lko_High"), color = "#56B4E9", active = T, query.name = "High CpGoe in 2"),

                  list(query = intersects, params = list("Foc_Low",  "Lko_Low"),  color = "#E69F00", active = T, query.name = "Low CpGoe in 2"),
                  list(query = intersects, params = list("Foc_Low",  "Lko_Low"),  color = "#E69F00", active = T, query.name = "Low CpGoe in 2"),
 
                  list(query = intersects, params = list("Gbi_Low",  "Lko_Low"),  color = "#E69F00", active = T, query.name = "Low CpGoe in 2"),
                  list(query = intersects, params = list("Foc_Low",  "Ame_Low"),  color = "#E69F00", active = T, query.name = "Low CpGoe in 2"),
                  list(query = intersects, params = list("Ame_Low",  "Gbi_Low"),  color = "#E69F00", active = T, query.name = "Low CpGoe in 2"),

                  list(query = intersects, params = list("Gbi_Low",  "Lko_Low",  "Ame_Low"),  color = "#E69F00", active = T, query.name = "Low CpGoe in  3"), 
                  list(query = intersects, params = list("Foc_Low",  "Gbi_Low",  "Ame_Low"),  color = "#E69F00", active = T, query.name = "Low CpGoe in  3"), 
                  list(query = intersects, params = list("Foc_Low",  "Lko_Low",  "Ame_Low"),  color = "#E69F00", active = T, query.name = "Low CpGoe in  3"), 

                   list(query = intersects, params = list("Gbi_Low",  "Lko_Low",  "Foc_Low"),  color = "#E69F00", active = T, query.name = "Low CpGoe in  3"), 
                  list(query = intersects, params = list("Gbi_High", "Lko_High", "Ame_High"), color = "#56B4E9", active = T, query.name = "High CpGoe in  3"),
                  list(query = intersects, params = list("Gbi_High", "Foc_High", "Ame_High"), color = "#56B4E9", active = T, query.name = "High CpGoe in  3"),
                  list(query = intersects, params = list("Gbi_High", "Foc_High", "Lko_High"), color = "#56B4E9", active = T, query.name = "High CpGoe in  3"),
                  list(query = intersects, params = list("Foc_High", "Ame_High", "Lko_High"), color = "#56B4E9", active = T, query.name = "High CpGoe in  3"),

                  list(query = intersects, params = list("Gbi_Low",  "Lko_Low",  "Ame_Low","Foc_Low"),  color = "#E69F00", active = T, query.name = "Low CpGoe in all 4"), 
                  list(query = intersects, params = list("Gbi_High", "Lko_High", "Ame_High","Foc_High"), color = "#56B4E9", active = T, query.name = "High CpGoe in all 4")
                 # list(query = intersects, params = list("Gbi_High", "Lko_High", "Ame_High","Foc_High","Gbi_Low",  "Lko_Low",  "Ame_Low","Foc_Low"), color = "red", active = T, query.name = "IN ALL")
                 ),
       query.legend = "bottom")
Upset_raw_v3


```


```{r}
Data_for_Upset_list_v3<-Data_for_Upset_list_v2
colnames(Data_for_Upset_list_v3)<-paste0(colnames(Data_for_Upset_list_v2), "_CpGoe")


Data_for_Upset_list_v4<- Data_for_Upset_list_v3 %>% 
  filter(rowSums(.)>3)
  

Upset_SubSetstoshow_v2<-upset(Data_for_Upset_list_v4,  nsets = 8, order.by = "freq", nintersects = 3,cutoff=10,               
                              sets = c("Foc_High_CpGoe","Ame_High_CpGoe","Lko_High_CpGoe", "Gbi_High_CpGoe","Foc_Low_CpGoe","Ame_Low_CpGoe", "Lko_Low_CpGoe","Gbi_Low_CpGoe"  ),keep.order = TRUE,
                  queries = list(
                  list(query = intersects, params = list("Gbi_Low_CpGoe",  "Lko_Low_CpGoe",  "Ame_Low_CpGoe","Foc_Low_CpGoe"),  color = "#E69F00", active = T, query.name = "Low CpGoe in all 3"), 
                  list(query = intersects, params = list("Gbi_High_CpGoe", "Lko_High_CpGoe", "Ame_High_CpGoe","Foc_High_CpGoe"), color = "#56B4E9", active = T, query.name = "High CpGoe in all 3")
        ),
       query.legend = "bottom")


Upset_SubSetstoshow_v2


```



## CpGoe Functional Analysis

- Gene orthology Enrichment Analysis with TopGO.

```{r, prepareGO, warning=FALSE, cache=TRUE,  results=FALSE}
library("topGO")

## Gbi
conGbi <- dbConnect(RSQLite::SQLite(), "GBI_V3_Annotations.v2.db")
Gbi_Gene_GO <-dbFetch( dbSendQuery(conGbi, "SELECT GeneID,GOterms FROM genes"))
dbDisconnect(conGbi)

dim(Gbi_Gene_GO)
write.table(Gbi_Gene_GO, "Gbi_Gene_GO.csv", row.names = FALSE, sep="\t", quote=FALSE)
geneID2GO_gbi <- readMappings("Gbi_Gene_GO.csv")
head(geneID2GO_gbi)

## Lko
conLko <- dbConnect(RSQLite::SQLite(), "Lko_Annotations.v2.db")
Lko_Gene_GO <-dbFetch( dbSendQuery(conLko, "SELECT GeneID,GOterms FROM genes"))
dbDisconnect(conLko)

dim(Lko_Gene_GO)
write.table(Lko_Gene_GO, "Lko_Gene_GO.csv", row.names = FALSE, sep="\t", quote=FALSE)
geneID2GO_lko <- readMappings("Lko_Gene_GO.csv")
head(geneID2GO_lko)

## Ame
### For Ame, I ran Iprscan to longest prot per geneql
head(Ame_CpG)
geneID2GO_Ame <- readMappings("Ame_Gene_GO_dict.txt")
head(geneID2GO_Ame)

## Foc
##./interproscan-5.34-73.0/interproscan.sh -appl pfam -dp -f TSV -goterms -iprlookup -pa -t p -i Frankliniella_occidentalis_longest_isoform_nodots.fa  -o Foc_longest_isoform.iprscan25/09/2020 14:46:29:735 Welcome to InterProScan-5.34-73.0
# python Interproscan_to_Gene_GOterm_Foc.py
geneID2GO_Foc <- readMappings("Foc_Gene_GO_dict.txt")

```


### *Gryllus bimaculatus* CpGoe GO-ernichment

I select genes belonging to the low CpGoe peak (methylated genes). And eprform a GOI-enrichment vs all genes.

#### Gbi Low CpGoe


```{r, GOBP_Gbi,  results=TRUE, cache=TRUE, message=F, warning=F}
## Low CpG
dim(Gbi_CpG)#68 less than the 17871, because NA & 0>0CpGoe<0 removed


GBi_LowCPG<-Gbi_CpG[Gbi_CpG$CpGoe_level=="Low", "Gene" ]


## selected genes/universe
geneListGbiLow=factor(as.integer(Gbi_CpG$Gene %in%  GBi_LowCPG))  ## being or not in the selected ones as Factor
names(geneListGbiLow)= Gbi_CpG$Gene

Gbi_Low_GOdataBP <- new("topGOdata",
                    description = "Gbi Low CpGoe", ontology = "BP",
                    allGenes =geneListGbiLow, 
                    ##geneSel = Gbi_CpG$Gene,
                    nodeSize = 10,
                    annot = annFUN.gene2GO,
                    gene2GO =geneID2GO_gbi)

Gbi_Low_GOdataBP

Gbi_Low_weight01_fisher_BP=runTest(Gbi_Low_GOdataBP, algorithm='weight01', statistic='fisher') 
#Gbi_Low_weight01_fisher_BP

allGO=usedGO(Gbi_Low_GOdataBP)

res_Gbi_Low_weight0_fisher_BP=GenTable(Gbi_Low_GOdataBP, weightFisher=Gbi_Low_weight01_fisher_BP, orderBy='weightFisher', topNodes=length(allGO), numChar=1000)
res_Gbi_Low_weight0_fisher_BP$pvalue<-topGO::score(Gbi_Low_weight01_fisher_BP)[res_Gbi_Low_weight0_fisher_BP$GO.ID]

#res_Gbi_Low_weight0_fisher_BP$weightFisher<-as.numeric(res_Gbi_Low_weight0_fisher_BP$weightFisher)
#head(res_Gbi_Low_weight0_fisher_BP)

#write.csv(res_Gbi_Low_weight0_fisher_BP, "res_Gbi_Low_weight0_fisher_BP")

```


#### Gbi High CpGoe


```{r, GOBP_Gbi_high, results=TRUE, cache=TRUE, message=F, warning=F}

# High CpG
GBi_HighCPG<-Gbi_CpG[Gbi_CpG$CpGoe_level=="High", "Gene" ]


## selected genes/universe
geneListGbiHigh=factor(as.integer(Gbi_CpG$Gene %in%   GBi_HighCPG))  ## being or not in the s;leected ones as Factor
names(geneListGbiHigh)= Gbi_CpG$Gene




Gbi_High_GOdataBP <- new("topGOdata",
                    description = "Gbi High CpGoe", ontology = "BP",
                    allGenes =geneListGbiHigh, 
                   # #geneSel = Gbi_CpG$Gene,
                    nodeSize = 10,
                    annot = annFUN.gene2GO,
                    gene2GO =geneID2GO_gbi)

Gbi_High_GOdataBP


Gbi_High_weight01_fisher_BP=runTest(Gbi_High_GOdataBP, algorithm='weight01', statistic='fisher') 
Gbi_High_weight01_fisher_BP

allGO=usedGO(Gbi_High_GOdataBP)


res_Gbi_High_weight0_fisher_BP<-GenTable(Gbi_High_GOdataBP, weightFisher=Gbi_High_weight01_fisher_BP, orderBy='weightFisher', topNodes=length(allGO), numChar=1000)
res_Gbi_High_weight0_fisher_BP$pvalue<-topGO::score(Gbi_High_weight01_fisher_BP)[res_Gbi_High_weight0_fisher_BP$GO.ID]



```

### *Laupala kohalensis* CpGoe GO-ernichment

#### Lko Low CpGoe

```{r, GOBP_Lko, results=TRUE, cache=TRUE, message=F, warning=F}

## Lko Low CpG
Lko_LowCPG<-Lko_CpG[Lko_CpG$CpGoe_level=="Low", "Gene" ]
length(Lko_LowCPG)
table(Lko_CpG$CpGoe_level)


## selected genes/universe
geneListLkoLow=factor(as.integer(Lko_CpG$Gene %in%  Lko_LowCPG))  ## being or not in the s;leected ones as Factor
names(geneListLkoLow)= Lko_CpG$Gene

                  
Lko_Low_GOdataBP <- new("topGOdata",
                    description = "Lko Low CpGoe", ontology = "BP",
                    allGenes =geneListLkoLow, 
                    #geneSel = Lko_CpG$Gene,
                    nodeSize = 10,
                    annot = annFUN.gene2GO,
                    gene2GO =geneID2GO_lko)

Lko_Low_GOdataBP


Lko_Low_weight01_fisher_BP=runTest(Lko_Low_GOdataBP, algorithm='weight01', statistic='fisher') 

allGO=usedGO(Lko_Low_GOdataBP)

res_Lko_Low_weight0_fisher_BP=GenTable(Lko_Low_GOdataBP, weightFisher=Lko_Low_weight01_fisher_BP, orderBy='weightFisher', topNodes=length(allGO), numChar=1000)
res_Lko_Low_weight0_fisher_BP$pvalue<-topGO::score(Lko_Low_weight01_fisher_BP)[res_Lko_Low_weight0_fisher_BP$GO.ID]




```

#### Lko High CpGoe

```{r, GOBP_Lko_high,  results=TRUE, cache=TRUE, message=F, warning=F}

# High CpG
Lko_HighCPG<-Lko_CpG[Lko_CpG$CpGoe_level=="High", "Gene" ]


## selected genes/universe
geneListLkoHigh=factor(as.integer(Lko_CpG$Gene %in%   Lko_HighCPG))  ## being or not in the s;leected ones as Factor
names(geneListLkoHigh)= Lko_CpG$Gene



Lko_High_GOdataBP <- new("topGOdata",
                    description = "Lko High CpGoe", ontology = "BP",
                    allGenes =geneListLkoHigh, 
                    #geneSel = Lko_CpG$Gene,
                    nodeSize = 10,
                    annot = annFUN.gene2GO,
                    gene2GO =geneID2GO_lko)

Lko_High_GOdataBP


Lko_High_weight01_fisher_BP=runTest(Lko_High_GOdataBP, algorithm='weight01', statistic='fisher') 
Lko_High_weight01_fisher_BP

allGO=usedGO(Lko_High_GOdataBP)


res_Lko_High_weight0_fisher_BP=GenTable(Lko_High_GOdataBP, weightFisher=Lko_High_weight01_fisher_BP, orderBy='weightFisher', topNodes=length(allGO), numChar=1000)
res_Lko_High_weight0_fisher_BP$pvalue<-topGO::score(Lko_High_weight01_fisher_BP)[res_Lko_High_weight0_fisher_BP$GO.ID]

```


### *Apis mellifera* CpGoe GO-ernichment

#### Ame Low CpGoe

```{r, GOBP_AmeLow, results=TRUE, cache=TRUE, message=F, warning=F}

Ame_cds_gene_dic<-read.delim("Ame_Gene_CDS_dict.txt", header = FALSE, sep=" ")
#head(Ame_cds_gene_dic)
colnames(Ame_cds_gene_dic)<-c("LOC","XM")
Ame_CpG_LOC<-merge(Ame_CpG,Ame_cds_gene_dic, by.x="Gene", by.y="XM" )


## Ame Low CpG
Ame_LowCPG<-Ame_CpG_LOC[Ame_CpG_LOC$CpGoe_level=="Low", "LOC" ]


## selected genes/universe
geneListAmeLow=factor(as.integer(Ame_CpG_LOC$LOC %in%  Ame_LowCPG))  ## being or not in the s;leected ones as Factor
names(geneListAmeLow)= Ame_CpG_LOC$LOC


Ame_Low_GOdataBP <- new("topGOdata",
                    description = "Ame Low CpGoe", ontology = "BP",
                    allGenes =geneListAmeLow, 
                    #geneSel = Ame_CpG$Gene,
                    nodeSize = 10,
                    annot = annFUN.gene2GO,
                    gene2GO =geneID2GO_Ame)

Ame_Low_GOdataBP


Ame_Low_weight01_fisher_BP=runTest(Ame_Low_GOdataBP, algorithm='weight01', statistic='fisher') 

allGO=usedGO(Ame_Low_GOdataBP)

res_Ame_Low_weight0_fisher_BP=GenTable(Ame_Low_GOdataBP, weightFisher=Ame_Low_weight01_fisher_BP, orderBy='weightFisher', topNodes=length(allGO), numChar=1000)

res_Ame_Low_weight0_fisher_BP$pvalue<-topGO::score(Ame_Low_weight01_fisher_BP)[res_Ame_Low_weight0_fisher_BP$GO.ID]


```

#### Ame High CpGoe

```{r, GOBP_AmeHigh, results=TRUE, cache=TRUE, message=F, warning=F}
## Ame High CpG
Ame_HighCPG<-Ame_CpG_LOC[Ame_CpG_LOC$CpGoe_level=="High", "LOC" ]


## selected genes/universe
geneListAmeHigh=factor(as.integer(Ame_CpG_LOC$LOC %in%  Ame_HighCPG))  ## being or not in the s;leected ones as Factor
names(geneListAmeHigh)= Ame_CpG_LOC$LOC


Ame_High_GOdataBP <- new("topGOdata",
                    description = "Ame High CpGoe", ontology = "BP",
                    allGenes =geneListAmeHigh, 
                    #geneSel = Ame_CpG$Gene,
                    nodeSize = 10,
                    annot = annFUN.gene2GO,
                    gene2GO =geneID2GO_Ame)

Ame_High_GOdataBP


Ame_High_weight01_fisher_BP=runTest(Ame_High_GOdataBP, algorithm='weight01', statistic='fisher') 

allGO=usedGO(Ame_High_GOdataBP)

res_Ame_High_weight0_fisher_BP=GenTable(Ame_High_GOdataBP, weightFisher=Ame_High_weight01_fisher_BP, orderBy='weightFisher', topNodes=length(allGO), numChar=1000)
res_Ame_High_weight0_fisher_BP$pvalue<-topGO::score(Ame_High_weight01_fisher_BP)[res_Ame_High_weight0_fisher_BP$GO.ID]


```




### *Frankliniella occidentalis* CpGoe GO-ernichment

#### Foc Low CpGoe

```{r, GOBP_FocLow, results=TRUE, cache=TRUE, message=F, warning=F}
geneID2GO_Foc <- readMappings("Foc_Gene_GO_dict.txt")
head(geneID2GO_Foc)
length(geneID2GO_Foc)
     

## Foc Low CpG
Foc_LowCPG<-Foc_CpG[Foc_CpG$CpGoe_level=="Low", "Gene" ]



## selected genes/universe
geneListFocLow=factor(as.integer(Foc_CpG$Gene %in%  Foc_LowCPG))  ## being or not in the s;leected ones as Factor
names(geneListFocLow)= Foc_CpG$Gene


Foc_Low_GOdataBP <- new("topGOdata",
                    description = "Foc Low CpGoe", ontology = "BP",
                    allGenes =geneListFocLow, 
                    #geneSel = Foc_CpG$Gene,
                    nodeSize = 10,
                    annot = annFUN.gene2GO,
                    gene2GO =geneID2GO_Foc)

Foc_Low_GOdataBP


Foc_Low_weight01_fisher_BP=runTest(Foc_Low_GOdataBP, algorithm='weight01', statistic='fisher') 
#Foc_Low_weight01_fisher_BP

allGO=usedGO(Foc_Low_GOdataBP)

res_Foc_Low_weight0_fisher_BP=GenTable(Foc_Low_GOdataBP, weightFisher=Foc_Low_weight01_fisher_BP, orderBy='weightFisher', topNodes=length(allGO), numChar=1000)
res_Foc_Low_weight0_fisher_BP$pvalue<-topGO::score(Foc_Low_weight01_fisher_BP)[res_Foc_Low_weight0_fisher_BP$GO.ID]



```

#### Foc High CpGoe

```{r, GOBP_FocHigh, results=TRUE, cache=TRUE, message=F, warning=F}
## Foc High CpG
Foc_HighCPG<-Foc_CpG[Foc_CpG$CpGoe_level=="High", "Gene" ]


## selected genes/universe
geneListFocHigh=factor(as.integer(Foc_CpG$Gene %in%  Foc_HighCPG))  ## being or not in the s;leected ones as Factor
names(geneListFocHigh)= Foc_CpG$Gene
#head(geneListFocHigh)
#table(geneListFocHigh)



Foc_High_GOdataBP <- new("topGOdata",
                    description = "Foc High CpGoe", ontology = "BP",
                    allGenes =geneListFocHigh, 
                    #geneSel = Foc_CpG$Gene,
                    nodeSize = 10,
                    annot = annFUN.gene2GO,
                    gene2GO =geneID2GO_Foc)

Foc_High_GOdataBP


Foc_High_weight01_fisher_BP=runTest(Foc_High_GOdataBP, algorithm='weight01', statistic='fisher') 

allGO=usedGO(Foc_High_GOdataBP)

res_Foc_High_weight0_fisher_BP=GenTable(Foc_High_GOdataBP, weightFisher=Foc_High_weight01_fisher_BP, orderBy='weightFisher', topNodes=length(allGO), numChar=1000)

res_Foc_High_weight0_fisher_BP$pvalue<-topGO::score(Foc_High_weight01_fisher_BP)[res_Foc_High_weight0_fisher_BP$GO.ID]


```

### GO Visualization

```{r}
All_Goes<-rbind(cbind(res_Gbi_High_weight0_fisher_BP, Class="Gbi_High"),
                cbind(res_Lko_High_weight0_fisher_BP, Class="Lko_High"),
                cbind(res_Foc_High_weight0_fisher_BP, Class="Foc_High"),
                cbind(res_Ame_High_weight0_fisher_BP, Class="Ame_High"),
                cbind(res_Gbi_Low_weight0_fisher_BP , Class="Gbi_Low"),
                cbind(res_Lko_Low_weight0_fisher_BP, Class="Lko_Low"),
                cbind(res_Foc_Low_weight0_fisher_BP, Class="Foc_Low"),
                cbind(res_Ame_Low_weight0_fisher_BP, Class="Ame_Low")
              )
```




### All significant GOS (Fig S3)

```{r}
#OrderTerms<- All_Goes %>%filter(pvalue<0.01) %>%  group_by(GO.ID,Term, Class) %>% tally(name="num",sort=TRUE) %>% ungroup() %>% group_by(GO.ID,Term) %>%tally(name="Ocurrences",sort=TRUE)
OrderTerms<- All_Goes %>%filter(pvalue<0.05)%>%  separate(col=Class, "_",into=c("Spp","HighLow"),remove=F) %>%  group_by(GO.ID,Term, HighLow,Class) %>% tally(name="num",sort=TRUE) %>% ungroup() %>% group_by(GO.ID,Term,HighLow) %>%tally(name="Ocurrences",sort=TRUE)  

Significants<-All_Goes %>%filter(pvalue<0.05) %>% dplyr::select(GO.ID) %>% distinct(GO.ID)

All_Goes_Factors<-All_Goes%>% 
    filter(GO.ID %in% Significants$GO.ID) %>% 
    left_join(OrderTerms, by=c("GO.ID"="GO.ID")) %>% 
    mutate(Term_factors=as.factor(Term.x)) %>% 
    mutate(Term_factors=fct_reorder2(Term_factors, Ocurrences,HighLow, .desc = TRUE)) %>% 
    mutate(significancelevel=ifelse(as.numeric(pvalue) <0.001,"<0.001", pvalue)) %>% 
    mutate(significancelevel=ifelse(as.numeric(pvalue) <=0.05 & as.numeric(pvalue)>0.001,"[0.05,0.001)",significancelevel)) %>% 
    mutate(significancelevel=ifelse(as.numeric(pvalue) >0.05,">0.05", significancelevel )) %>% 
    mutate(Class=fct_relevel(Class, c("Gbi_Low","Lko_Low","Foc_Low","Ame_Low","Gbi_High","Lko_High","Foc_High","Ame_High") ) ) %>% 
   mutate(significancelevel=as.factor(significancelevel) ) %>% 
  mutate(significancelevel=fct_relevel(significancelevel, c(">0.05","[0.05,0.001)","<0.001") )  ) 


```

- Fig S3, too large to plot here

```{r}

# ggplot(All_Goes_Factors, aes(Term_factors, Class  )) + 
#   geom_point(aes(fill=significancelevel,  size = Significant/Annotated*100),shape = 21, stroke = 0.1)+
#   coord_flip()+
#   scale_fill_tableau(palette = "Tableau 10", type = c("regular"), direction = 1)+
#   theme_bw()+
#   labs(y="",x="")+
#      theme(title=element_text(size=10, hjust=0.5),
#         axis.title=element_text(size=10),
#         axis.text = element_text(size=10),
#         axis.text.x = element_text(angle = 45, hjust=1) )

```





### Selected GO terms to show in Fig 3

```{r}
library(ggthemes)

## Selected Goterms to show
TOPSignificants<-All_Goes %>%filter(pvalue<0.00001) %>% dplyr::select(GO.ID) %>% distinct(GO.ID)

Selected_Goes_Factors<-All_Goes_Factors %>% filter(GO.ID %in% TOPSignificants$GO.ID)


SELECTED_Point_plot<-ggplot(Selected_Goes_Factors, aes(Term_factors, Class  )) + 
  geom_point(aes(fill=significancelevel,  size = Significant/Annotated*100),shape = 21, stroke = 0.1)+
  coord_flip()+
  scale_fill_tableau(palette = "Tableau 10", type = c("regular"), direction = 1)+
  theme_bw()+
  labs(y="",x="")+
     theme(title=element_text(size=10, hjust=0.5),
        axis.title=element_text(size=10),
        axis.text = element_text(size=10),
        axis.text.x = element_text(angle = 45, hjust=1) )

SELECTED_Point_plot

```





## CpGoe vs dNds (Fig 2D)


- Load dNdS results from the dNds Naysis [file]("https://github.com/guillemylla/Gryllus_genome_annotation/tree/master/Comparative_Genomics_V2/dNdS")


```{r, loadDnDs, warning=FALSE}
Omega_Gbi_Lko<-read.csv("dNdS_Gbi_Lko.csv", header = T, sep = "\t", stringsAsFactors = FALSE)
Omega_Gbi_Lko$OG<-sapply(strsplit(as.character(Omega_Gbi_Lko$key),'_'), `[`, 2)

```

```{r, AddOGtoCpG, warning=FALSE, fig.width=12, fig.height=20}
Omega_Gbi_Lko_CpG0<-merge(Omega_Gbi_Lko, AmeGbiLko_CpGeo_OG[AmeGbiLko_CpGeo_OG$Abb=="Gbi",c("OG", "CpGoe","CpGoe_level")], by.x="OG", by.y="OG", all.y=FALSE)


Omega_Gbi_Lko_CpG<-merge(Omega_Gbi_Lko_CpG0, AmeGbiLko_CpGeo_OG[AmeGbiLko_CpGeo_OG$Abb=="Lko",c("OG", "CpGoe","CpGoe_level")], by.x="OG", by.y="OG",suffixes=c("Gbi","Lko"), all.y=FALSE)

Omega_Gbi_Lko_CpG$Both_CpGlevels<-paste(Omega_Gbi_Lko_CpG$CpGoe_levelGbi, Omega_Gbi_Lko_CpG$CpGoe_levelLko, sep="_")

prop.table(table(Omega_Gbi_Lko_CpG$Both_CpGlevels))*100


##Filter genes omega=none
Omega_Gbi_Lko_CpG<-Omega_Gbi_Lko_CpG[!Omega_Gbi_Lko_CpG$Omega %like% "None",]

Omega_Gbi_Lko_CpG$Both_CpGlevels <- as.factor(Omega_Gbi_Lko_CpG$Both_CpGlevels)

Omega_Gbi_Lko_CpG$Both_CpGlevels <- factor(Omega_Gbi_Lko_CpG$Both_CpGlevels,levels = c("Low_Low","High_High","High_Low" , "Low_High"))

```



```{r, PlotCpgDnDds1, warning=FALSE}

ggplot(subset(Omega_Gbi_Lko_CpG, Both_CpGlevels=="High_High" | Both_CpGlevels=="Low_Low" ), aes(x=Both_CpGlevels, y=as.numeric(Omega))) +
                        geom_violin()+
                        geom_boxplot(width=0.1) +
                        ggtitle("dN/dS (omega) Gbi-Lko")+
                        ylab("Omega (dD/dS)")+xlab("CpGo/e level in both insects")

```


```{r cpgdndstest}

## It applies Wilcoxon-Mann-Whitney test (Mann-Whitney one data is not paired)
wilcox.test(subset(Omega_Gbi_Lko_CpG, Both_CpGlevels=="High_High")$Omega, subset(Omega_Gbi_Lko_CpG, Both_CpGlevels=="Low_Low")$Omega, paired=F)
```